--- triton/BUILD	2023-03-10 07:24:02.000000000 -0800
+++ triton/BUILD	2023-03-22 14:25:49.000000000 -0700
@@ -442,6 +442,7 @@
         "//third_party/llvm/llvm-project/llvm:IRReader",
         "//third_party/llvm/llvm-project/llvm:Linker",
         "//third_party/llvm/llvm-project/llvm:Support",
+        "//third_party/llvm/llvm-project/mlir:BuiltinToLLVMIRTranslation",
         "//third_party/llvm/llvm-project/mlir:ConversionPasses",
         "//third_party/llvm/llvm-project/mlir:ExecutionEngine",
         "//third_party/llvm/llvm-project/mlir:ExecutionEngineUtils",
--- triton/lib/Target/LLVMIR/LLVMIRTranslation.cpp	2023-03-10 07:24:02.000000000 -0800
+++ triton/lib/Target/LLVMIR/LLVMIRTranslation.cpp	2023-03-22 14:25:49.000000000 -0700
@@ -7,6 +7,7 @@
 #include "mlir/IR/Dialect.h"
 #include "mlir/Pass/Pass.h"
 #include "mlir/Pass/PassManager.h"
+#include "mlir/Target/LLVMIR/Dialect/Builtin/BuiltinToLLVMIRTranslation.h"
 #include "mlir/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.h"
 #include "mlir/Target/LLVMIR/Dialect/NVVM/NVVMToLLVMIRTranslation.h"
 #include "mlir/Target/LLVMIR/Export.h"
@@ -233,6 +234,7 @@
 std::unique_ptr<llvm::Module>
 translateLLVMToLLVMIR(llvm::LLVMContext *llvmContext, mlir::ModuleOp module) {
   DialectRegistry registry;
+  mlir::registerBuiltinDialectTranslation(registry);
   mlir::registerLLVMDialectTranslation(registry);
   mlir::registerNVVMDialectTranslation(registry);
   module->getContext()->appendDialectRegistry(registry);
