==== triton/BUILD#9 - triton/BUILD#10 ====
# action=edit type=text
--- triton/BUILD	2023-03-10 07:24:02.000000000 -0800
+++ triton/BUILD	2023-03-23 08:38:30.000000000 -0700
@@ -442,6 +442,7 @@
         "@llvm-project//llvm:IRReader",
         "@llvm-project//llvm:Linker",
         "@llvm-project//llvm:Support",
+        "@llvm-project//mlir:BuiltinToLLVMIRTranslation",
         "@llvm-project//mlir:ConversionPasses",
         "@llvm-project//mlir:ExecutionEngine",
         "@llvm-project//mlir:ExecutionEngineUtils",
==== triton/lib/Target/LLVMIR/LLVMIRTranslation.cpp#9 - triton/lib/Target/LLVMIR/LLVMIRTranslation.cpp#10 ====
# action=edit type=text
--- triton/lib/Target/LLVMIR/LLVMIRTranslation.cpp	2023-03-10 07:24:02.000000000 -0800
+++ triton/lib/Target/LLVMIR/LLVMIRTranslation.cpp	2023-03-23 08:38:30.000000000 -0700
@@ -7,6 +7,7 @@
 #include "mlir/IR/Dialect.h"
 #include "mlir/Pass/Pass.h"
 #include "mlir/Pass/PassManager.h"
+#include "mlir/Target/LLVMIR/Dialect/Builtin/BuiltinToLLVMIRTranslation.h"
 #include "mlir/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.h"
 #include "mlir/Target/LLVMIR/Dialect/NVVM/NVVMToLLVMIRTranslation.h"
 #include "mlir/Target/LLVMIR/Export.h"
@@ -233,6 +234,7 @@
 std::unique_ptr<llvm::Module>
 translateLLVMToLLVMIR(llvm::LLVMContext *llvmContext, mlir::ModuleOp module) {
   DialectRegistry registry;
+  mlir::registerBuiltinDialectTranslation(registry);
   mlir::registerLLVMDialectTranslation(registry);
   mlir::registerNVVMDialectTranslation(registry);
   module->getContext()->appendDialectRegistry(registry);
