==== triton/BUILD#15 - triton/BUILD ====
# action=edit type=text
--- triton/BUILD	2023-04-24 19:41:15.000000000 -0700
+++ triton/BUILD	2023-06-08 04:26:58.000000000 -0700
@@ -37,6 +37,22 @@
 # exports_files(["LICENSE"])
 # copybara:uncomment_end
 
+config_setting(
+    name = "compiler_is_msvc",
+    flag_values = {
+        # copybara:comment_begin
+        "@bazel_tools" +
+        # copybara:comment_end
+        "//tools/cpp:compiler": "msvc-cl",
+    },
+)
+
+# TODO(csigg): fix, enable error upstream, remove.
+_no_unused_variable = select({
+    ":compiler_is_msvc": [],
+    "//conditions:default": ["-Wno-unused-variable"],
+})
+
 td_library(
     name = "td_files",
     srcs = glob(["include/triton/**/*.td"]),
@@ -276,7 +292,7 @@
     name = "TritonDialect",
     srcs = glob(["lib/Dialect/Triton/IR/*.cpp"]),
     hdrs = glob(["include/triton/Dialect/Triton/IR/*.h"]),
-    copts = ["-Wno-unused-variable"],  # TODO(manany): fix
+    copts = _no_unused_variable,
     includes = ["include"],
     deps = [
         ":triton_dialect_inc_gen",
@@ -328,7 +344,7 @@
         "include/triton/Analysis/*.h",
         "include/triton/Dialect/TritonGPU/IR/*.h",
     ]),
-    copts = ["-Wno-unused-variable"],  # TODO(csigg): fix
+    copts = _no_unused_variable,
     includes = ["include"],
     deps = [
         ":TritonDialect",
@@ -356,7 +372,7 @@
         "lib/Dialect/TritonGPU/Transforms/*.h",
     ]),
     hdrs = glob(["include/triton/Dialect/TritonGPU/Transforms/*.h"]),
-    copts = ["-Wno-unused-variable"],  # TODO(csigg): fix
+    copts = _no_unused_variable,
     includes = ["include"],
     deps = [
         ":TritonDialect",
@@ -391,7 +407,7 @@
         "include/triton/Tools/Sys/*.hpp",
         "include/triton/Conversion/TritonGPUToLLVM/*.h",
     ]),
-    copts = ["-Wno-unused-variable"],  # TODO(csigg): fix
+    copts = _no_unused_variable,
     includes = [
         "include",
         "lib/Conversion/TritonGPUToLLVM",
