diff --git triton/lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp triton/lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp
index 32605e605..07901f06f 100644
--- triton/lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp
+++ triton/lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp
@@ -333,7 +333,8 @@ public:
 
     mlir::RewritePatternSet patterns(context);
     patterns.add<ConvertTransConvert>(context);
-    patterns.add<MoveOpAfterLayoutConversion>(context);
+    // TODO(b/283035396): Fix CUDA_ERROR_MISALIGNED_ADDRESS and uncomment.
+    // patterns.add<MoveOpAfterLayoutConversion>(context);
     if (applyPatternsAndFoldGreedily(m, std::move(patterns)).failed())
       signalPassFailure();
     if (fixupLoops(m).failed())

