diff --git a/BUILD b/BUILD
index a5a813485..c7f8aa5a6 100644
--- a/BUILD
+++ b/BUILD
@@ -275,8 +275,7 @@ cc_library(
     copts = _no_unused_variable,
     includes = ["include"],
     deps = [
-        ":TritonDialect",
-        ":TritonGPUDialect",
+        ":TritonDialects",
         ":TritonTools",
         ":triton_gpu_attr_inc_gen",
         "@llvm-project//llvm:Support",
@@ -291,44 +290,53 @@ cc_library(
 )
 
 cc_library(
-    name = "TritonDialect",
-    srcs = glob(["lib/Dialect/Triton/IR/*.cpp"]),
-    hdrs = glob(["include/triton/Dialect/Triton/IR/*.h"]),
+    name = "TritonDialects",
+    srcs = glob([
+        "lib/Dialect/Triton/IR/*.cpp",
+        "lib/Dialect/TritonGPU/IR/*.cpp",
+    ]) + [
+        "include/triton/Analysis/Utility.h",  # Avoid circular dependency.
+    ],
+    hdrs = glob([
+        "include/triton/Dialect/Triton/IR/*.h",
+        "include/triton/Dialect/TritonGPU/IR/*.h",
+    ]),
     copts = _no_unused_variable,
     includes = ["include"],
     deps = [
-        ":TritonGPUAttributes",
         ":triton_dialect_inc_gen",
+        ":triton_gpu_attr_inc_gen",
+        ":triton_gpu_dialect_inc_gen",
+        ":triton_gpu_ops_inc_gen",
+        ":triton_gpu_transforms_inc_gen",
         ":triton_interfaces_inc_gen",
         ":triton_ops_inc_gen",
         "@llvm-project//llvm:Support",
+        "@llvm-project//mlir:Analysis",
         "@llvm-project//mlir:ArithDialect",
         "@llvm-project//mlir:ControlFlowDialect",
         "@llvm-project//mlir:ControlFlowInterfaces",
+        "@llvm-project//mlir:DestinationStyleOpInterface",
         "@llvm-project//mlir:FuncDialect",
+        "@llvm-project//mlir:GPUDialect",
         "@llvm-project//mlir:IR",
+        "@llvm-project//mlir:LLVMDialect",
         "@llvm-project//mlir:MathDialect",
         "@llvm-project//mlir:Pass",
         "@llvm-project//mlir:SCFDialect",
         "@llvm-project//mlir:Support",
         "@llvm-project//mlir:TensorDialect",
+        "@llvm-project//mlir:Transforms",
     ],
 )
 
-cc_library(
-    name = "TritonGPUAttributes",
-    hdrs = ["include/triton/Dialect/TritonGPU/IR/Attributes.h"],
-    includes = ["include"],
-    deps = ["triton_gpu_attr_inc_gen"],
-)
-
 cc_library(
     name = "TritonTransforms",
     srcs = glob(["lib/Dialect/Triton/Transforms/*.cpp"]),
     hdrs = glob(["include/triton/Dialect/Triton/Transforms/*.h"]),
     includes = ["include"],
     deps = [
-        ":TritonDialect",
+        ":TritonDialects",
         ":triton_combine_inc_gen",
         ":triton_transforms_inc_gen",
         "@llvm-project//llvm:Support",
@@ -347,36 +355,6 @@ cc_library(
     alwayslink = True,  # TritonDialect uses getCanonicalizationPatterns().
 )
 
-cc_library(
-    name = "TritonGPUDialect",
-    srcs = glob(["lib/Dialect/TritonGPU/IR/*.cpp"]),
-    hdrs = [
-        "include/triton/Analysis/Utility.h",  # Avoid circular dependency.
-        "include/triton/Dialect/TritonGPU/IR/Dialect.h",
-        "include/triton/Dialect/TritonGPU/IR/Traits.h",
-    ],
-    copts = _no_unused_variable,
-    includes = ["include"],
-    deps = [
-        ":TritonDialect",
-        ":TritonGPUAttributes",
-        ":triton_gpu_attr_inc_gen",
-        ":triton_gpu_dialect_inc_gen",
-        ":triton_gpu_ops_inc_gen",
-        ":triton_gpu_transforms_inc_gen",
-        "@llvm-project//llvm:Support",
-        "@llvm-project//mlir:Analysis",
-        "@llvm-project//mlir:DestinationStyleOpInterface",
-        "@llvm-project//mlir:GPUDialect",
-        "@llvm-project//mlir:IR",
-        "@llvm-project//mlir:LLVMDialect",
-        "@llvm-project//mlir:Pass",
-        "@llvm-project//mlir:Support",
-        "@llvm-project//mlir:TensorDialect",
-        "@llvm-project//mlir:Transforms",
-    ],
-)
-
 cc_library(
     name = "TritonGPUTransforms",
     srcs = glob([
@@ -388,9 +366,7 @@ cc_library(
     includes = ["include"],
     deps = [
         ":TritonAnalysis",
-        ":TritonDialect",
-        ":TritonGPUAttributes",
-        ":TritonGPUDialect",
+        ":TritonDialects",
         ":triton_gpu_transforms_inc_gen",
         "@llvm-project//llvm:Support",
         "@llvm-project//mlir:Analysis",
@@ -428,8 +404,7 @@ cc_library(
     ],
     deps = [
         ":TritonAnalysis",
-        ":TritonDialect",
-        ":TritonGPUDialect",
+        ":TritonDialects",
         ":triton_conversion_triton_gpu_to_llvm_passes_inc_gen",
         ":triton_conversion_triton_to_triton_gpu_passes_inc_gen",
         "@llvm-project//llvm:Support",
@@ -466,8 +441,8 @@ cc_library(
     hdrs = glob(["include/triton/Conversion/TritonToTritonGPU/*.h"]),
     includes = ["include"],
     deps = [
-        ":TritonDialect",
-        ":TritonGPUDialect",
+        ":TritonAnalysis",
+        ":TritonDialects",
         ":TritonGPUTransforms",
         ":triton_conversion_triton_gpu_to_llvm_passes_inc_gen",
         ":triton_conversion_triton_to_triton_gpu_passes_inc_gen",
@@ -513,9 +488,7 @@ cc_library(
         "@llvm-project//mlir:ROCDLToLLVMIRTranslation",
         "@llvm-project//mlir:ToLLVMIRTranslation",
         "@llvm-project//mlir:Transforms",
-        # copybara:uncomment_begin
-        # "//third_party/py/triton/google:find_cuda",
-        # copybara:uncomment_end
+        # copybara:uncomment "//third_party/py/triton/google:find_cuda",
     ],
 )
 
@@ -579,8 +552,7 @@ cc_binary(
     ],
     includes = ["include"],
     deps = [
-        ":TritonDialect",
-        ":TritonGPUDialect",
+        ":TritonDialects",
         ":TritonGPUToLLVM",
         ":TritonGPUTransforms",
         ":TritonToTritonGPU",
@@ -618,8 +590,7 @@ cc_binary(
     ],
     includes = ["include"],
     deps = [
-        ":TritonDialect",
-        ":TritonGPUDialect",
+        ":TritonDialects",
         ":TritonGPUToLLVM",
         ":TritonGPUTransforms",
         ":TritonHSACO",
