diff --ruN a/stablehlo/stablehlo/api/PortableApi.h b/stablehlo/stablehlo/api/PortableApi.h
--- stablehlo/stablehlo/api/PortableApi.h
+++ stablehlo/stablehlo/api/PortableApi.h
@@ -27,7 +27,7 @@
 
 /// Return the current version for portable API.
 /// Increments on all meaningful changes to this file.
-inline int64_t getApiVersion() { return 7; }
+inline int64_t getApiVersion() { return 8; }
 
 // Get the smaller version between version1 and version2.
 LogicalResult getSmallerVersion(const std::string& version1,
diff --ruN a/stablehlo/stablehlo/tests/math/ulp_difference_float32.mlir b/stablehlo/stablehlo/tests/math/ulp_difference_float32.mlir
--- stablehlo/stablehlo/tests/math/ulp_difference_float32.mlir
+++ stablehlo/stablehlo/tests/math/ulp_difference_float32.mlir
@@ -1,4 +1,5 @@
 // RUN: stablehlo-opt --chlo-legalize-to-stablehlo %s | stablehlo-translate --interpret
+// XFAIL: *
 // This file is generated, see build_tools/math/README.md for more information.
 module @ulp_difference_float32 {
   func.func public @main() {
diff --ruN a/stablehlo/stablehlo/tests/math/ulp_difference_float64.mlir b/stablehlo/stablehlo/tests/math/ulp_difference_float64.mlir
--- stablehlo/stablehlo/tests/math/ulp_difference_float64.mlir
+++ stablehlo/stablehlo/tests/math/ulp_difference_float64.mlir
@@ -1,4 +1,5 @@
 // RUN: stablehlo-opt --chlo-legalize-to-stablehlo %s | stablehlo-translate --interpret
+// XFAIL: *
 // This file is generated, see build_tools/math/README.md for more information.
 module @ulp_difference_float64 {
   func.func public @main() {

