diff --ruN a/stablehlo/stablehlo/dialect/StablehloOps.cpp b/stablehlo/stablehlo/dialect/StablehloOps.cpp
--- stablehlo/stablehlo/dialect/StablehloOps.cpp
+++ stablehlo/stablehlo/dialect/StablehloOps.cpp
@@ -253,9 +253,9 @@
     return failure();
 
   return verifyReduceScatter(*this,
-                             /*operand_types=*/{getOperand().getType()},
-                             /*result_types=*/{getType()},
-                             /*scatter_dimension=*/getScatterDimension());
+                             /*operandTypes=*/{getOperand().getType()},
+                             /*resultTypes=*/{getType()},
+                             /*scatterDimension=*/getScatterDimension());
 }
 
 //===----------------------------------------------------------------------===//
@@ -4361,7 +4361,7 @@
   if (failed(windowStridesOrErr)) return failure();
   auto windowOrErr = hlo::verifyWindowAttributesAndInferWindowDimensions(
       *windowDimsOrErr, *windowStridesOrErr, *paddingOrErr,
-      /*lhs_dilation=*/{}, /*rhs_dilation=*/{}, getLoc());
+      /*lhsDilation=*/{}, /*rhsDilation=*/{}, getLoc());
   if (failed(windowOrErr)) return failure();
 
   // P5.
diff --ruN a/stablehlo/stablehlo/dialect/TypeInference.cpp b/stablehlo/stablehlo/dialect/TypeInference.cpp
--- stablehlo/stablehlo/dialect/TypeInference.cpp
+++ stablehlo/stablehlo/dialect/TypeInference.cpp
@@ -867,8 +867,8 @@
   if (failed(windowDilationsOrErr)) return failure();
   auto windowOrErr = verifyWindowAttributesAndInferWindowDimensions(
       *windowDimsOrErr, *windowStridesOrErr, *paddingOrErr,
-      /*lhs_dilation=*/*baseDilationsOrErr,
-      /*rhs_dilation=*/*windowDilationsOrErr, location);
+      /*lhsDilation=*/*baseDilationsOrErr,
+      /*rhsDilation=*/*windowDilationsOrErr, location);
   if (failed(windowOrErr)) return failure();
 
   // P5.

