diff --ruN a/stablehlo/BUILD.bazel b/stablehlo/BUILD.bazel
--- stablehlo/BUILD.bazel
+++ stablehlo/BUILD.bazel
@@ -278,6 +278,7 @@
         ":chlo_attrs_inc_gen",
         ":chlo_enums_inc_gen",
         ":chlo_ops_inc_gen",
+        ":stablehlo_type_inference",
         "@llvm-project//llvm:Support",
         "@llvm-project//mlir:BytecodeReader",
         "@llvm-project//mlir:BytecodeWriter",
@@ -813,7 +814,9 @@
         "@llvm-project//mlir:AllPassesAndDialects",
         "@llvm-project//mlir:FuncDialect",
         "@llvm-project//mlir:IR",
-        "@llvm-project//mlir:Support",
+        "@llvm-project//mlir:Pass",
+        "@llvm-project//mlir:Support",
+        "@llvm-project//mlir:Transforms",
         "@llvm-project//mlir:TranslateLib",
     ],
 )
diff --ruN a/stablehlo/stablehlo/dialect/CMakeLists.txt b/stablehlo/stablehlo/dialect/CMakeLists.txt
--- stablehlo/stablehlo/dialect/CMakeLists.txt
+++ stablehlo/stablehlo/dialect/CMakeLists.txt
@@ -65,6 +65,7 @@
   LINK_LIBS PUBLIC
   StablehloBase
   StablehloBroadcastUtils
+  StablehloTypeInference
   MLIRArithDialect
   MLIRControlFlowInterfaces
   MLIRInferTypeOpInterface
diff --ruN a/stablehlo/stablehlo/dialect/ChloOps.cpp b/stablehlo/stablehlo/dialect/ChloOps.cpp
--- stablehlo/stablehlo/dialect/ChloOps.cpp
+++ stablehlo/stablehlo/dialect/ChloOps.cpp
@@ -28,6 +28,7 @@
 #include "mlir/Transforms/InliningUtils.h"
 #include "stablehlo/dialect/BroadcastUtils.h"
 #include "stablehlo/dialect/ChloBytecode.h"
+#include "stablehlo/dialect/TypeInference.h"
 
 // Include order matters
 #include "stablehlo/dialect/ChloEnums.cpp.inc"
@@ -453,30 +454,9 @@
     ValueShapeRange operands, DictionaryAttr attributes,
     OpaqueProperties properties, RegionRange regions,
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  Builder builder(context);
   TopKOp::Adaptor adaptor(operands, attributes, properties, regions);
-  Value operand = adaptor.getOperand();
-  uint64_t k = adaptor.getK();
-
-  auto operandTy = operand.getType().dyn_cast<RankedTensorType>();
-  if (!operandTy) return emitOptionalError(location, "operand must be ranked");
-  if (operandTy.getRank() < 1)
-    return emitOptionalError(location, "operand's rank must be at least 1");
-  auto operandLastDim = operandTy.getShape()[operandTy.getRank() - 1];
-  if (operandLastDim == ShapedType::kDynamic)
-    return emitOptionalError(location,
-                             "operand's last dimension must be static");
-  if (operandLastDim < static_cast<int64_t>(k))
-    return emitOptionalError(location,
-                             "operand's last dimension must be at least ", k);
-
-  SmallVector<int64_t> resultShape;
-  append_range(resultShape, operandTy.getShape());
-  resultShape[operandTy.getRank() - 1] = k;
-
-  inferredReturnShapes.emplace_back(resultShape, operandTy.getElementType());
-  inferredReturnShapes.emplace_back(resultShape, builder.getI32Type());
-  return success();
+  return hlo::inferTopKOp(location, adaptor.getOperand(), adaptor.getK(),
+                          inferredReturnShapes);
 }
 
 //===----------------------------------------------------------------------===//
diff --ruN a/stablehlo/stablehlo/dialect/TypeInference.cpp b/stablehlo/stablehlo/dialect/TypeInference.cpp
--- stablehlo/stablehlo/dialect/TypeInference.cpp
+++ stablehlo/stablehlo/dialect/TypeInference.cpp
@@ -2854,6 +2854,42 @@
   return success();
 }
 
+LogicalResult inferTopKOp(
+    std::optional<Location> location, Value operand, int64_t k,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  Builder builder(operand.getContext());
+  auto operandType = operand.getType().dyn_cast<RankedTensorType>();
+  if (!operandType) {
+    inferredReturnShapes.emplace_back(
+        operand.getType().cast<ShapedType>().getElementType());
+    inferredReturnShapes.emplace_back(builder.getI32Type());
+    return success();
+  }
+
+  if (operandType.getRank() < 1)
+    return emitOptionalError(location, "operand's rank must be at least 1");
+  auto operandLastDim = operandType.getRank() - 1;
+  if (!operandType.isDynamicDim(operandLastDim) &&
+      operandType.getDimSize(operandLastDim) < k)
+    return emitOptionalError(location,
+                             "operand's last dimension must be at least ", k);
+
+  SmallVector<int64_t> resultShape(operandType.getShape());
+  resultShape[operandLastDim] = k;
+  SmallVector<int64_t> resultBounds(
+      encodingToBounds(operandType.getEncoding()));
+  if (!resultBounds.empty())
+    resultBounds[operandLastDim] = ShapedType::kDynamic;
+
+  inferredReturnShapes.emplace_back(
+      resultShape, operandType.getElementType(),
+      hlo::boundsToEncoding(operandType.getEncoding(), resultBounds));
+  inferredReturnShapes.emplace_back(
+      resultShape, builder.getI32Type(),
+      hlo::boundsToEncoding(operandType.getEncoding(), resultBounds));
+  return success();
+}
+
 LogicalResult inferTransposeOp(std::optional<Location> loc, Value operand,
                                DenseIntElementsAttr permutation,
                                SmallVectorImpl<Type>& inferredReturnTypes) {
diff --ruN a/stablehlo/stablehlo/dialect/TypeInference.h b/stablehlo/stablehlo/dialect/TypeInference.h
--- stablehlo/stablehlo/dialect/TypeInference.h
+++ stablehlo/stablehlo/dialect/TypeInference.h
@@ -335,6 +335,10 @@
     std::optional<Location> location, ValueRange inputs,
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
 
+LogicalResult inferTopKOp(
+    std::optional<Location> location, Value operand, int64_t k,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
+
 LogicalResult inferTransposeOp(std::optional<Location> loc, Value operand,
                                DenseIntElementsAttr permutation,
                                SmallVectorImpl<Type>& inferredReturnTypes);
diff --ruN a/stablehlo/stablehlo/tests/ops_chlo.mlir b/stablehlo/stablehlo/tests/ops_chlo.mlir
--- stablehlo/stablehlo/tests/ops_chlo.mlir
+++ stablehlo/stablehlo/tests/ops_chlo.mlir
@@ -90,28 +90,10 @@
 
 // -----
 
-func.func @top_k(%arg0 : tensor<*xf32>) {
-  // expected-error @+2 {{failed to infer returned types}}
-  // @expected-error @+1{{operand must be ranked}}
-  %0:2 = chlo.top_k(%arg0, k=8) : tensor<*xf32> -> (tensor<8xf32>, tensor<8xi32>)
-  return
-}
-
-// -----
-
 func.func @top_k(%arg0 : tensor<f32>) {
   // expected-error @+2 {{failed to infer returned types}}
   // @expected-error @+1{{operand's rank must be at least 1}}
   %0:2 = chlo.top_k(%arg0, k=8) : tensor<f32> -> (tensor<8xf32>, tensor<8xi32>)
-  return
-}
-
-// -----
-
-func.func @top_k(%arg0 : tensor<?xf32>) {
-  // expected-error @+2 {{failed to infer returned types}}
-  // @expected-error @+1{{operand's last dimension must be static}}
-  %0:2 = chlo.top_k(%arg0, k=8) : tensor<?xf32> -> (tensor<8xf32>, tensor<8xi32>)
   return
 }
 
@@ -126,15 +108,36 @@
 
 // -----
 
-func.func @top_k(%arg0 : tensor<16xf32>) {
+func.func @top_k_1d(%arg0 : tensor<16xf32>) {
   %0:2 = chlo.top_k(%arg0, k=8) : tensor<16xf32> -> (tensor<8xf32>, tensor<8xi32>)
   return
 }
 
 // -----
 
-func.func @top_k(%arg0 : tensor<16x16xf32>) {
+func.func @top_k_nd(%arg0 : tensor<16x16xf32>) {
   %0:2 = chlo.top_k(%arg0, k=8) : tensor<16x16xf32> -> (tensor<16x8xf32>, tensor<16x8xi32>)
+  return
+}
+
+// -----
+
+func.func @top_k_unbounded(%arg0 : tensor<?x16x?xf32>) {
+  %0:2 = chlo.top_k(%arg0, k=8) : tensor<?x16x?xf32> -> (tensor<?x16x8xf32>, tensor<?x16x8xi32>)
+  return
+}
+
+// -----
+
+func.func @top_k_bounded(%arg0 : tensor<?x?x?xf32, #stablehlo.bounds<?, 16, 16>>) {
+  %0:2 = chlo.top_k(%arg0, k=8) : tensor<?x?x?xf32, #stablehlo.bounds<?, 16, 16>> -> (tensor<16x?x8xf32, #stablehlo.bounds<?, 16, ?>>, tensor<16x?x8xi32, #stablehlo.bounds<?, 16, ?>>)
+  return
+}
+
+// -----
+
+func.func @top_k_unranked(%arg0 : tensor<*xf32>) {
+  %0:2 = chlo.top_k(%arg0, k=8) : tensor<*xf32> -> (tensor<*xf32>, tensor<*xi32>)
   return
 }
 
diff --ruN a/stablehlo/stablehlo/tools/CMakeLists.txt b/stablehlo/stablehlo/tools/CMakeLists.txt
--- stablehlo/stablehlo/tools/CMakeLists.txt
+++ stablehlo/stablehlo/tools/CMakeLists.txt
@@ -44,7 +44,9 @@
   ${dialect_libs}
   MLIRFuncDialect
   MLIRIR
+  MLIRPass
   MLIRSupport
+  MLIRTransforms
   MLIRTranslateLib
   CheckOps
   StablehloOps

