diff --git a/.gitignore b/.gitignore
index 25bf90ad2f..7572e592d0 100644
--- a/.gitignore
+++ b/.gitignore
@@ -83,8 +83,6 @@ hooks/
 localhooks/
 
 # created when packaging, don't version control this
-src/scip/githash.c
-src/scip/buildflags.c
 
 # settings
 settings/
diff --git a/src/scip/benders_xyz.c b/src/scip/benders_xyz.c
index d39b973c51..19e8540cd0 100644
--- a/src/scip/benders_xyz.c
+++ b/src/scip/benders_xyz.c
@@ -56,6 +56,7 @@
 /** Benders' decomposition data */
 struct SCIP_BendersData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/benderscut_xyz.c b/src/scip/benderscut_xyz.c
index 5eb41adbe9..281dd29a9a 100644
--- a/src/scip/benderscut_xyz.c
+++ b/src/scip/benderscut_xyz.c
@@ -50,6 +50,7 @@
 /** Benders' decomposition cut data */
 struct SCIP_BenderscutData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/branch_xyz.c b/src/scip/branch_xyz.c
index b0f6d9528a..7ed4f817bf 100644
--- a/src/scip/branch_xyz.c
+++ b/src/scip/branch_xyz.c
@@ -51,6 +51,7 @@
 /** branching rule data */
 struct SCIP_BranchruleData
 {
+  void* ptr;
 };
 
 
@@ -225,7 +226,7 @@ SCIP_RETCODE SCIPincludeBranchruleXyz(
    /* use SCIPincludeBranchrule() if you want to set all callbacks explicitly and realize (by getting compiler errors) when
     * new callbacks are added in future SCIP versions
     */
-   SCIP_CALL( SCIPincludeBranchrule(scip, BRANCHRULE_NAME, BRANCHRULE_DESC, BRANCHRULE_PRIORITY, BRANCHRULE_MAXDEPTH, 
+   SCIP_CALL( SCIPincludeBranchrule(scip, BRANCHRULE_NAME, BRANCHRULE_DESC, BRANCHRULE_PRIORITY, BRANCHRULE_MAXDEPTH,
          BRANCHRULE_MAXBOUNDDIST,
          branchCopyXyz, branchFreeXyz, branchInitXyz, branchExitXyz, branchInitsolXyz, branchExitsolXyz,
          branchExeclpXyz, branchExecextXyz, branchExecpsXyz,
diff --git a/src/scip/compr_xyz.c b/src/scip/compr_xyz.c
index 348d61c75c..322c2446cf 100644
--- a/src/scip/compr_xyz.c
+++ b/src/scip/compr_xyz.c
@@ -50,6 +50,7 @@
 /** tree compression data */
 struct SCIP_ComprData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/cons_xyz.c b/src/scip/cons_xyz.c
index d4b750f2ec..8178cdc5b7 100644
--- a/src/scip/cons_xyz.c
+++ b/src/scip/cons_xyz.c
@@ -78,11 +78,13 @@
 /** constraint data for xyz constraints */
 struct SCIP_ConsData
 {
+  void* ptr;
 };
 
 /** constraint handler data */
 struct SCIP_ConshdlrData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/cutsel_xyz.c b/src/scip/cutsel_xyz.c
index e78da4eb4a..6203ed00a5 100644
--- a/src/scip/cutsel_xyz.c
+++ b/src/scip/cutsel_xyz.c
@@ -49,6 +49,7 @@
 /** cut selector data */
 struct SCIP_CutselData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/dialog_xyz.c b/src/scip/dialog_xyz.c
index 05354614ae..866269e490 100644
--- a/src/scip/dialog_xyz.c
+++ b/src/scip/dialog_xyz.c
@@ -37,7 +37,7 @@
 
 #define DIALOG_NAME            "xyz"
 #define DIALOG_DESC            "xyz user interface dialog"
-#define DIALOG_ISSUBMENU          FALSE 
+#define DIALOG_ISSUBMENU          FALSE
 
 
 
@@ -51,6 +51,7 @@
 /** dialog data */
 struct SCIP_DialogData
 {
+  void* ptr;
 };
 
 
@@ -163,7 +164,7 @@ SCIP_RETCODE SCIPincludeDialogXyz(
    /* create, include, and release dialog */
    if( !SCIPdialogHasEntry(parentdialog, DIALOG_NAME) )
    {
-      SCIP_CALL( SCIPincludeDialog(scip, &dialog, 
+      SCIP_CALL( SCIPincludeDialog(scip, &dialog,
             dialogCopyXyz, dialogExecXyz, dialogDescXyz, dialogFreeXyz,
             DIALOG_NAME, DIALOG_DESC, DIALOG_ISSUBMENU, dialogdata) );
       SCIP_CALL( SCIPaddDialogEntry(scip, parentdialog, dialog) );
diff --git a/src/scip/disp_xyz.c b/src/scip/disp_xyz.c
index ac1d7ccef9..09f71c67cd 100644
--- a/src/scip/disp_xyz.c
+++ b/src/scip/disp_xyz.c
@@ -35,13 +35,13 @@
 #include "scip/disp_xyz.h"
 
 
-#define DISP_NAME               "xyz"                
+#define DISP_NAME               "xyz"
 #define DISP_DESC               "xyz display column"
-#define DISP_HEADER             "xyz" 
+#define DISP_HEADER             "xyz"
 #define DISP_WIDTH              14      /**< the width of the display column */
 #define DISP_PRIORITY           110000  /**< the priority of the display column */
 #define DISP_POSITION           30100   /**< the relative position of the display column */
-#define DISP_STRIPLINE          TRUE    /**< the default for whether the display column should be separated 
+#define DISP_STRIPLINE          TRUE    /**< the default for whether the display column should be separated
                                          *   with a line from its right neighbor */
 
 
@@ -56,6 +56,7 @@
 /** display column data */
 struct SCIP_DispData
 {
+  void* ptr;
 };
 
 
@@ -197,10 +198,10 @@ SCIP_RETCODE SCIPincludeDispXyz(
    /* TODO: (optional) create display column specific data here */
 
    /* include display column */
-   SCIP_CALL( SCIPincludeDisp(scip, DISP_NAME, DISP_DESC, DISP_HEADER, SCIP_DISPSTATUS_AUTO, 
+   SCIP_CALL( SCIPincludeDisp(scip, DISP_NAME, DISP_DESC, DISP_HEADER, SCIP_DISPSTATUS_AUTO,
          dispCopyXyz,
-         dispFreeXyz, dispInitXyz, dispExitXyz, 
-         dispInitsolXyz, dispExitsolXyz, dispOutputXyz, 
+         dispFreeXyz, dispInitXyz, dispExitXyz,
+         dispInitsolXyz, dispExitsolXyz, dispOutputXyz,
          dispdata, DISP_WIDTH, DISP_PRIORITY, DISP_POSITION, DISP_STRIPLINE) );
 
    /* add xyz display column parameters */
diff --git a/src/scip/event_xyz.c b/src/scip/event_xyz.c
index 7d63318d6c..e878ff7062 100644
--- a/src/scip/event_xyz.c
+++ b/src/scip/event_xyz.c
@@ -45,6 +45,7 @@
 /** event handler data */
 struct SCIP_EventhdlrData
 {
+  void* ptr;
 };
 
 /*
@@ -188,7 +189,7 @@ SCIP_RETCODE SCIPincludeEventHdlrXyz(
     */
    SCIP_CALL( SCIPincludeEventhdlr(scip, EVENTHDLR_NAME, EVENTHDLR_DESC,
          eventCopyXyz,
-         eventFreeXyz, eventInitXyz, eventExitXyz, 
+         eventFreeXyz, eventInitXyz, eventExitXyz,
          eventInitsolXyz, eventExitsolXyz, eventDeleteXyz, eventExecXyz,
          eventhdlrdata) );
 #else
diff --git a/src/scip/expr_xyz.c b/src/scip/expr_xyz.c
index 63b0638b78..39101dd0ec 100644
--- a/src/scip/expr_xyz.c
+++ b/src/scip/expr_xyz.c
@@ -47,11 +47,13 @@
 /** expression handler data */
 struct SCIP_ExprhdlrData
 {
+  void* ptr;
 };
 
 /** expression data */
 struct SCIP_ExprData
 {
+  void* ptr;
 };
 
 /*
diff --git a/src/scip/githash.c b/src/scip/githash.c
new file mode 100644
index 0000000000..943de15fa4
--- /dev/null
+++ b/src/scip/githash.c
@@ -0,0 +1 @@
+#define SCIP_GITHASH "62fab8a2e3"
diff --git a/src/scip/heur_xyz.c b/src/scip/heur_xyz.c
index 519a55d98f..22944d128b 100644
--- a/src/scip/heur_xyz.c
+++ b/src/scip/heur_xyz.c
@@ -55,6 +55,7 @@
 /** primal heuristic data */
 struct SCIP_HeurData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/nlhdlr_xyz.c b/src/scip/nlhdlr_xyz.c
index f459ba5f05..70f82dde0e 100644
--- a/src/scip/nlhdlr_xyz.c
+++ b/src/scip/nlhdlr_xyz.c
@@ -49,11 +49,13 @@
 /** nonlinear handler data */
 struct SCIP_NlhdlrData
 {
+  void* ptr;
 };
 
 /** nonlinear handler expression data */
 struct SCIP_NlhdlrExprData
 {
+  void* ptr;
 };
 
 /*
diff --git a/src/scip/nlpi_xyz.c b/src/scip/nlpi_xyz.c
index 54866878fd..7212ca0aba 100644
--- a/src/scip/nlpi_xyz.c
+++ b/src/scip/nlpi_xyz.c
@@ -52,12 +52,14 @@
 
 struct SCIP_NlpiData
 {
+  void* ptr;
 };
 
 /* TODO: fill in the necessary NLP problem instance data */
 
 struct SCIP_NlpiProblem
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/nodesel_xyz.c b/src/scip/nodesel_xyz.c
index 054500369c..938f1c26ac 100644
--- a/src/scip/nodesel_xyz.c
+++ b/src/scip/nodesel_xyz.c
@@ -50,6 +50,7 @@
 /** node selector data */
 struct SCIP_NodeselData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/presol_xyz.c b/src/scip/presol_xyz.c
index 82e1a1a60a..4be46acda7 100644
--- a/src/scip/presol_xyz.c
+++ b/src/scip/presol_xyz.c
@@ -51,6 +51,7 @@
 /** presolver data */
 struct SCIP_PresolData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/pricer_xyz.c b/src/scip/pricer_xyz.c
index b21be1c1c7..086c7b1c1e 100644
--- a/src/scip/pricer_xyz.c
+++ b/src/scip/pricer_xyz.c
@@ -52,6 +52,7 @@
 /** variable pricer data */
 struct SCIP_PricerData
 {
+  void* ptr;
 };
 
 
@@ -213,7 +214,7 @@ SCIP_RETCODE SCIPincludePricerXyz(
     * new callbacks are added in future SCIP versions
     */
    SCIP_CALL( SCIPincludePricer(scip, PRICER_NAME, PRICER_DESC, PRICER_PRIORITY, PRICER_DELAY,
-         pricerCopyXyz, pricerFreeXyz, pricerInitXyz, pricerExitXyz, 
+         pricerCopyXyz, pricerFreeXyz, pricerInitXyz, pricerExitXyz,
          pricerInitsolXyz, pricerExitsolXyz, pricerRedcostXyz, pricerFarkasXyz,
          pricerdata) );
 #else
diff --git a/src/scip/prop_xyz.c b/src/scip/prop_xyz.c
index 2f5151463c..91415c5aa9 100644
--- a/src/scip/prop_xyz.c
+++ b/src/scip/prop_xyz.c
@@ -37,7 +37,7 @@
 /* fundamental propagator properties */
 #define PROP_NAME              "xyz"
 #define PROP_DESC              "propagator template"
-#define PROP_PRIORITY                 0 /**< propagator priority */ 
+#define PROP_PRIORITY                 0 /**< propagator priority */
 #define PROP_FREQ                    10 /**< propagator frequency */
 #define PROP_DELAY                FALSE /**< should propagation method be delayed, if other propagators found reductions? */
 #define PROP_TIMING             SCIP_PROPTIMING_BEFORELP/**< propagation timing mask */
@@ -59,6 +59,7 @@
 /** propagator data */
 struct SCIP_PropData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/reader_xyz.c b/src/scip/reader_xyz.c
index c1c538794c..6c984717aa 100644
--- a/src/scip/reader_xyz.c
+++ b/src/scip/reader_xyz.c
@@ -49,6 +49,7 @@
 /** data for xyz reader */
 struct SCIP_ReaderData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/relax_xyz.c b/src/scip/relax_xyz.c
index efdfa41a88..64bca3d139 100644
--- a/src/scip/relax_xyz.c
+++ b/src/scip/relax_xyz.c
@@ -52,6 +52,7 @@
 /** relaxator data */
 struct SCIP_RelaxData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/scipbuildflags.c b/src/scip/scipbuildflags.c
index 27ab6082a2..e2d6a7caec 100644
--- a/src/scip/scipbuildflags.c
+++ b/src/scip/scipbuildflags.c
@@ -30,10 +30,9 @@
 
 /*---+----1----+----2----+----3----+----4----+----5----+----6----+----7----+----8----+----9----+----0----+----1----+----2*/
 
+#define SCIP_BUILDFLAGS " ARCH=x86_64\n COMP=gnu\n DEBUGSOL=false\n EXPRINT=none\n GAMS=false\n SYM=bliss\n GMP=false\n IPOPT=false\n IPOPTOPT=opt\n WORHP=false\n WORHPOPT=opt\n LPS=spx2\n LPSCHECK=false\n LPSOPT=opt\n NOBLKBUFMEM=false\n NOBLKMEM=false\n NOBUFMEM=false\n OPT=opt\n OSTYPE=linux\n PARASCIP=true\n READLINE=false\n SANITIZE=\n SHARED=false\n USRARFLAGS=\n USRCFLAGS=-fPIC\n USRCXXFLAGS=-fPIC\n USRDFLAGS=\n USRFLAGS=\n USRLDFLAGS=\n USROFLAGS=\n VERSION=7.0.1\n ZIMPL=false\n ZIMPLOPT=opt\n ZLIB=true"
+
 #include "scip/scipbuildflags.h"
-#ifdef NO_CONFIG_HEADER
-#include "buildflags.c"
-#endif
 
 /** returns the flags that were used to build SCIP */
 const char* SCIPgetBuildFlags(
diff --git a/src/scip/sepa_xyz.c b/src/scip/sepa_xyz.c
index 75f3f810a8..e1d8171a48 100644
--- a/src/scip/sepa_xyz.c
+++ b/src/scip/sepa_xyz.c
@@ -53,6 +53,7 @@
 /** separator data */
 struct SCIP_SepaData
 {
+  void* ptr;
 };
 
 
diff --git a/src/scip/table_xyz.c b/src/scip/table_xyz.c
index 62d596871b..42fb15d3a8 100644
--- a/src/scip/table_xyz.c
+++ b/src/scip/table_xyz.c
@@ -52,6 +52,7 @@
 /** statistics table data */
 struct SCIP_TableData
 {
+  void* ptr;
 };
 
 
diff --git a/src/symmetry/compute_symmetry_bliss.cpp b/src/symmetry/compute_symmetry_bliss.cpp
index 203a64a384..82e0725aa1 100644
--- a/src/symmetry/compute_symmetry_bliss.cpp
+++ b/src/symmetry/compute_symmetry_bliss.cpp
@@ -34,8 +34,8 @@
 #include "compute_symmetry.h"
 
 /* include bliss graph */
-#include <bliss/defs.hh>
-#include <bliss/graph.hh>
+#include <bliss-0.73/defs.hh>
+#include <bliss-0.73/graph.hh>
 
 #include <string.h>
 #include <vector>
