
--- a/lib/Conversion/TritonGPUToLLVM/ControlFlowOpToLLVM.cpp	2024-03-19 09:23:43.000000000 -0700
+++ b/lib/Conversion/TritonGPUToLLVM/ControlFlowOpToLLVM.cpp	2024-10-01 02:58:18.000000000 -0700
@@ -104,9 +104,26 @@
                 this->getTypeConverter()->packFunctionResults(resultTypes)))
         return nullptr;
     }
+    // Add LLVMOp Bundle Attrs
+    // https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/CodeGen/CodeGen.cpp#L113-L131
+    llvm::SmallVector<mlir::NamedAttribute> newAttrs;
+    newAttrs.reserve(callOp->getAttrs().size() + 2);
+
+    for (mlir::NamedAttribute attr : callOp->getAttrs()) {
+      if (attr.getName() != "operandSegmentSizes")
+        newAttrs.push_back(attr);
+    }
+
+    newAttrs.push_back(rewriter.getNamedAttr(
+        "operandSegmentSizes",
+        rewriter.getDenseI32ArrayAttr(
+            {static_cast<int>(promotedOperands.size()), 0})));
+    newAttrs.push_back(rewriter.getNamedAttr(
+        "op_bundle_sizes", rewriter.getDenseI32ArrayAttr({})));
+
     auto newCallOp = rewriter.create<LLVM::CallOp>(
         callOp.getLoc(), packedResult ? TypeRange(packedResult) : TypeRange(),
-        promotedOperands, callOp->getAttrs());
+        promotedOperands, newAttrs);
     return newCallOp;
   }
 

--- a/third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp	2024-09-25 10:13:59.000000000 -0700
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp	2024-09-30 23:51:44.000000000 -0700
@@ -190,7 +190,8 @@
       auto name = StringAttr::get(callOp.getContext(), "llvm.amdgcn.rcp.f32");
       LLVM::FastmathFlagsAttr defaultFlags{};
       auto rcpOp = rewriter.create<LLVM::CallIntrinsicOp>(
-          loc, returnType, name, operands[1], defaultFlags);
+          loc, returnType, name, operands[1], defaultFlags,
+          ::llvm::ArrayRef<::mlir::ValueRange>{} /*op_bundle_operands*/);
 
       replacementOp = rewriter.create<LLVM::FMulOp>(
           loc, returnType, operands[0], rcpOp->getResult(0), defaultFlags);


--- a/third_party/amd/lib/TritonAMDGPUToLLVM/DotOpToLLVM/WMMA.cpp	2024-08-20 03:28:55.000000000 -0700
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/DotOpToLLVM/WMMA.cpp	2024-09-30 23:51:44.000000000 -0700
@@ -219,7 +219,8 @@
   }
   auto wmmaIntrinsic = rewriter.create<mlir::LLVM::CallIntrinsicOp>(
       loc, TypeRange{valC.getType()}, StringAttr::get(loc.getContext(), name),
-      operands, defaultFlags);
+      operands, defaultFlags,
+      ::llvm::ArrayRef<::mlir::ValueRange>{} /*op_bundle_operands*/);
 
   return wmmaIntrinsic.getResult(0);
 }


--- a/third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp	2024-09-16 13:44:40.000000000 -0700
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp	2024-09-30 23:51:44.000000000 -0700
@@ -72,7 +72,10 @@
   auto stringAttr = rewriter.getStringAttr("llvm.amdgcn.ballot");
   SmallVector<Value> operands = {cmp};
   Value asmResult =
-      rewriter.create<LLVM::CallIntrinsicOp>(loc, type, stringAttr, operands)
+      rewriter
+          .create<LLVM::CallIntrinsicOp>(
+              loc, type, stringAttr, operands, ::mlir::LLVM::FastmathFlags{},
+              ::llvm::ArrayRef<::mlir::ValueRange>{} /*op_bundle_operands*/)
           ->getResult(0);
   return asmResult;
 }


--- a/third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp
@@ -48,9 +48,10 @@ void createSchedGroupBarrier(PatternRewr
                                           static_cast<int32_t>(groupIdValue));
 
   LLVM::FastmathFlagsAttr defaultFlags{};
-  rewriter.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
-                                         ValueRange{mask, size, groupId},
-                                         defaultFlags);
+  rewriter.create<LLVM::CallIntrinsicOp>(
+      loc, TypeRange{}, intrinsicName, ValueRange{mask, size, groupId},
+      defaultFlags,
+      ::llvm::ArrayRef<::mlir::ValueRange>{} /*op_bundle_operands*/);
 }
 
 // Insert intrinsic that controls the types of instructions that may be
@@ -63,8 +64,9 @@ Operation *createSchedBarrier(PatternRew
 
   Value mask =
       LLVM::createConstantI32(loc, rewriter, static_cast<int32_t>(maskValue));
-  return rewriter.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
-                                                ValueRange{mask}, defaultFlags);
+  return rewriter.create<LLVM::CallIntrinsicOp>(
+      loc, TypeRange{}, intrinsicName, ValueRange{mask}, defaultFlags,
+      ::llvm::ArrayRef<::mlir::ValueRange>{} /*op_bundle_operands*/);
 }
 
 // Insert an experimental intrinsic for instruction group level parallelism.
@@ -76,7 +78,8 @@ Operation *createIglpOpt(PatternRewriter
   Value iglpValue =
       LLVM::createConstantI32(loc, rewriter, static_cast<int32_t>(value));
   return rewriter.create<LLVM::CallIntrinsicOp>(
-      loc, TypeRange{}, intrinsicName, ValueRange{iglpValue}, defaultFlags);
+      loc, TypeRange{}, intrinsicName, ValueRange{iglpValue}, defaultFlags,
+      ::llvm::ArrayRef<::mlir::ValueRange>{} /*op_bundle_operands*/);
 }
 
 struct InstructionSchedHintsRewriter
