
--- a/third_party/amd/lib/TritonAMDGPUToLLVM/BufferOpsEmitter.cpp	2024-11-07 04:49:10.000000000 -0800
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/BufferOpsEmitter.cpp	2025-01-23 08:33:43.000000000 -0800
@@ -1,10 +1,10 @@
-#include "PatternTritonGPUOpToLLVM.h"
-#include "TargetInfo.h"
-#include "Utility.h"
 #include "mlir/Dialect/LLVMIR/LLVMTypes.h"
 #include "mlir/Dialect/LLVMIR/NVVMDialect.h"
 #include "mlir/Dialect/LLVMIR/ROCDLDialect.h"
 #include "mlir/IR/PatternMatch.h"
+#include "PatternTritonGPUOpToLLVM.h"
+#include "TargetInfo.h"
+#include "Utility.h"
 #include "triton/Conversion/TritonGPUToLLVM/TypeConverter.h"
 #include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 #include "triton/Dialect/Triton/IR/Dialect.h"

--- a/third_party/nvidia/BUILD	2025-01-23 03:06:03.000000000 -0800
+++ b/third_party/nvidia/BUILD	2025-01-23 08:24:44.000000000 -0800
@@ -84,6 +84,7 @@
         "@llvm-project//mlir:Support",
         "@llvm-project//mlir:TransformUtils",
         "@llvm-project//mlir:Transforms",
+        "//:TritonAnalysis",
         "//:TritonDialects",
     ],
 )

--- a/third_party/nvidia/lib/NVGPUToLLVM/NVGPUToLLVMPass.cpp	2025-01-21 05:40:49.000000000 -0800
+++ b/third_party/nvidia/lib/NVGPUToLLVM/NVGPUToLLVMPass.cpp	2025-01-23 08:24:44.000000000 -0800
@@ -10,6 +10,7 @@
 
 #include "nvidia/lib/TritonNVIDIAGPUToLLVM/Utility.h"
 #include "llvm/Support/ErrorHandling.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 
 using namespace mlir;
 using namespace mlir::triton;

--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2OrV3.cpp	2025-01-21 05:40:49.000000000 -0800
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2OrV3.cpp	2025-01-23 08:24:46.000000000 -0800
@@ -2,6 +2,7 @@
 #include "Utility.h"
 #include "mlir/Support/LLVM.h"
 #include "third_party/nvidia/include/Dialect/NVGPU/IR/Dialect.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 
 using namespace mlir;
 

--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/DotOpToLLVM.cpp	2024-12-03 10:55:13.000000000 -0800
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/DotOpToLLVM.cpp	2025-01-23 08:24:45.000000000 -0800
@@ -2,6 +2,7 @@
 #include "Utility.h"
 
 #include "triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 
 using namespace mlir;
 using namespace mlir::triton;

--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/DotOpToLLVM/WGMMA.cpp	2025-01-22 04:23:33.000000000 -0800
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/DotOpToLLVM/WGMMA.cpp	2025-01-23 08:24:46.000000000 -0800
@@ -23,6 +23,7 @@
 
 #include "Utility.h"
 #include "mlir/Support/LLVM.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 
 using namespace mlir;
 using namespace mlir::triton;

--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TargetInfo.cpp	2024-12-05 23:53:31.000000000 -0800
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TargetInfo.cpp	2025-01-23 08:24:45.000000000 -0800
@@ -6,6 +6,7 @@
 #include "mlir/Dialect/LLVMIR/LLVMTypes.h"
 #include "mlir/Dialect/LLVMIR/NVVMDialect.h"
 #include "llvm/Support/MathExtras.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 
 using namespace mlir;
 

--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TensorPtrOpsToLLVM.cpp	2024-03-11 11:42:57.000000000 -0700
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TensorPtrOpsToLLVM.cpp	2025-01-23 08:24:45.000000000 -0800
@@ -23,6 +23,7 @@
 #include "PatternTritonGPUOpToLLVM.h"
 #include "Utility.h"
 #include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 
 using namespace mlir;
 using namespace mlir::triton;

--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/Utility.cpp	2024-06-27 02:04:25.000000000 -0700
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/Utility.cpp	2025-01-23 08:24:45.000000000 -0800
@@ -2,6 +2,7 @@
 #include "Dialect/NVGPU/IR/Dialect.h"
 #include "mlir/Dialect/LLVMIR/NVVMDialect.h"
 #include "triton/Conversion/TritonGPUToLLVM/TypeConverter.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 
 namespace mlir {
 namespace LLVM {

--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/Utility.h	2024-07-09 08:04:00.000000000 -0700
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/Utility.h	2025-01-23 08:24:45.000000000 -0800
@@ -3,8 +3,6 @@
 
 #include "nvidia/include/TritonNVIDIAGPUToLLVM/PTXAsmFormat.h"
 
-#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
-
 #include "mlir/Conversion/LLVMCommon/Pattern.h"
 #include "mlir/Dialect/LLVMIR/LLVMDialect.h"
 #include "third_party/nvidia/include/Dialect/NVGPU/IR/Dialect.h"
