
--- a/third_party/nvidia/lib/Dialect/NVWS/Transforms/InsertTmemAref.cpp	2025-09-25 06:36:50.000000000 -0700
+++ b/third_party/nvidia/lib/Dialect/NVWS/Transforms/InsertTmemAref.cpp	2025-09-29 04:55:20.000000000 -0700
@@ -54,8 +54,8 @@
     SmallVector<std::unique_ptr<Node>> subDags;
     Node(Operation *op, OpOperand *tokOperand,
          std::optional<PartitionId> partitionId, Node *parent)
-        : op(op), tokOperand(tokOperand), partitionId(partitionId),
-          parent(parent), parentDag(nullptr) {}
+        : op(op), tokOperand(tokOperand), parent(parent), parentDag(nullptr),
+          partitionId(partitionId) {}
 
     // ------------------------------------------------------------------------
 
@@ -364,7 +364,7 @@
   enum Kind { PUT, GET };
 
   TMEMAref(Value aref, Value origBuffer, Value replToken)
-      : aref(aref), origBuffer(origBuffer), replToken(replToken), kind(PUT) {}
+      : origBuffer(origBuffer), aref(aref), replToken(replToken), kind(PUT) {}
 
   void acquire(OpBuilder &b, Location loc,
                std::pair<std::optional<PartitionId>, StageCluster>
