https://github.com/triton-lang/triton/pull/9059
--- a/third_party/amd/include/Utils/Utility.h
+++ b/third_party/amd/include/Utils/Utility.h
@@ -1,7 +1,7 @@
 #ifndef TRITON_THIRD_PARTY_AMD_INCLUDE_UTILS_UTILITY_H_
 #define TRITON_THIRD_PARTY_AMD_INCLUDE_UTILS_UTILITY_H_
 
-#include "llvm/ADT/ArrayRef.h"
+#include "mlir/Support/LLVM.h"
 #include <cassert>
 #include <vector>
 namespace mlir::LLVM::AMD {
https://github.com/triton-lang/triton/pull/9060
--- a/third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp
+++ a/third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp
@@ -105,7 +105,7 @@
 
   explicit TritonAMDGPULowerInstructionSchedHints(StringRef arch,
                                                   int32_t numStages) {
-    this->arch = std::move(arch.str());
+    this->arch = arch.str();
     this->numStages = numStages;
   }
 
@@ -138,7 +138,7 @@
           TritonAMDGPUInsertInstructionSchedHints> {
 
   explicit TritonAMDGPUInsertInstructionSchedHints(StringRef variant) {
-    this->variant = std::move(variant.str());
+    this->variant = variant.str();
   }
 
   void runOnOperation() override {
