
--- a/python/test/BUILD	2026-02-10 05:37:32.000000000 -0800
+++ b/python/test/BUILD	2026-02-12 04:24:42.000000000 -0800
@@ -276,7 +276,7 @@
     exec_properties = _asan_needs_extra_link_memory,
     tags = [
         "config-cuda-only",
-        "requires-gpu-sm90:2",
+        "requires-gpu-sm90-full:2",
     ],
     target_compatible_with = _requires_config_cuda,
     tests = ["unit/runtime/test_peer_access.py"],
