This should be merged with the launcher patch afterwards.

--- a/third_party/nvidia/backend/cuda_utils.cc	2025-11-18 05:00:39.000000000 -0800
+++ b/third_party/nvidia/backend/cuda_utils.cc	2025-12-02 08:10:32.000000000 -0800
@@ -180,6 +180,14 @@
     clusterDimSchedulingAttr.value.clusterSchedulingPolicyPreference =
         CU_CLUSTER_SCHEDULING_POLICY_SPREAD;
   }
+
+  // As per the comment in third_party/nvidia/backend/driver.py,
+  // "num_ctas == 16 is non-portable. Does work for H100 and B200 tho".
+  if (config.num_ctas == 16) {
+    CUDA_CHECK(cuFuncSetAttribute(
+        config.function, CU_FUNC_ATTRIBUTE_NON_PORTABLE_CLUSTER_SIZE_ALLOWED,
+        1));
+  }
 
   // cuLaunchKernelEx was added in CUDA 12, so load it dynamically to be
   // able to link on CUDA 11 and earlier.
