Merge with launcher.patch

--- a/third_party/nvidia/backend/cuda_utils.cc
+++ b/third_party/nvidia/backend/cuda_utils.cc
@@ -605,8 +605,8 @@
     return nullptr;
   }
 
-  // +1 for the global scratch pointer.
-  std::size_t num_params = signature_metadata.size() + 1;
+  // +2 for the global scratch pointer and profile scratch pointer.
+  std::size_t num_params = signature_metadata.size() + 2;
   // Use alloca to set up kernel parameters on the stack and avoid dynamic
   // memory allocations.
   config.params = static_cast<void**>(alloca(num_params * sizeof(void*)));
