
--- a/third_party/nvidia/backend/cuda_utils.cc	2025-11-13 05:31:00.000000000 -0800
+++ b/third_party/nvidia/backend/cuda_utils.cc	2025-11-14 02:50:50.000000000 -0800
@@ -59,7 +59,7 @@
 #define CUDA_CHECK_AND_RETURN_NULL(ans)                                        \
   do {                                                                         \
     if (!gpuAssert((ans), __FILE__, __LINE__))                                 \
-      goto cleanup;                                                            \
+      return NULL;                                                            \
   } while (0)
 
 // To be used inside a Py_{BEGIN,END}_ALLOW_THREADS block.
@@ -77,7 +77,7 @@
     if ((funcPointer) == NULL) {                                               \
       (funcPointer) = (initializerFunction)();                                 \
       if ((funcPointer) == NULL) {                                             \
-        goto cleanup;                                                          \
+        return NULL;                                                          \
       }                                                                        \
     }                                                                          \
   } while (0)
@@ -912,16 +912,21 @@
 
 // clang-format off
 static PyTypeObject PyCUtensorMapType = {
-    PyVarObject_HEAD_INIT(NULL, 0)
+    .ob_base = {
+        .ob_base = {
+            .ob_type = NULL,
+        },
+        .ob_size = 0,
+    },
     .tp_name = "triton.backends.nvidia.PyCUtensorMap",
     .tp_basicsize = sizeof(PyCUtensorMapObject),
     .tp_itemsize = 0,
+    .tp_dealloc = (destructor)PyCUtensorMap_dealloc,
     .tp_flags = Py_TPFLAGS_DEFAULT,
     .tp_doc = "<PyCUtensorMap object>",
     .tp_methods = PyCUtensorMap_methods,
+    .tp_alloc = PyCUtensorMap_alloc,
     .tp_new = PyType_GenericNew,
-    .tp_alloc = PyCUtensorMap_alloc,
-    .tp_dealloc = (destructor)PyCUtensorMap_dealloc,
     .tp_free = PyCUtensorMap_free,
 };
 // clang-format on
@@ -1056,9 +1061,11 @@
   INITIALIZE_FUNCTION_POINTER_IF_NULL(cuTensorMapEncodeTiled,
                                       getCuTensorMapEncodeTiledHandle);
   CUresult res = cuTensorMapEncodeTiled(
-      &desc->tensorMap, elemType, rank, (void *)global_address, shapeInt,
-      stridesLL, blockSizeInt, elementStrides, CU_TENSOR_MAP_INTERLEAVE_NONE,
-      swizzle, CU_TENSOR_MAP_L2_PROMOTION_L2_128B, fill);
+      &desc->tensorMap, (CUtensorMapDataType)elemType, rank,
+      (void*)global_address, shapeInt, stridesLL, blockSizeInt, elementStrides,
+      CU_TENSOR_MAP_INTERLEAVE_NONE, (CUtensorMapSwizzle)swizzle,
+      CU_TENSOR_MAP_L2_PROMOTION_L2_128B, fill);
+
   if (res != CUDA_SUCCESS) {
     const char *str;
     cuGetErrorString(res, &str);
@@ -1104,8 +1111,6 @@
   }
 
   return (PyObject *)desc;
-
-  return result;
 }
 
 static PyMethodDef ModuleMethods[] = {
