--- a/python/test/unit/language/test_core.py
+++ b/python/test/unit/language/test_core.py
@@ -363,8 +363,7 @@ def _test_binary(dtype_x, dtype_y, expr,
             # We remove any explicit casting
             pattern = r'\.astype\(np\.\w+\)'
             scalar_expr = expr if numpy_expr is None else re.sub(pattern, '', numpy_expr)
-            with promotion_numpy_2_0():
-                z_ref = eval(scalar_expr)
+            z_ref = eval(scalar_expr)
         else:
             z_ref = eval(expr if numpy_expr is None else numpy_expr)
 
