diff --git a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp
--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp
@@ -1088,6 +1088,11 @@ struct AsyncCopyGlobalToLocalOpConversio
     //     mask bits are the same.  For example if N=2, the mask must be
     //     [x, x, y, y, ...].
     unsigned maxVec = getContiguity(op.getSrc());
+
+    if (resElemTy.isF16() && maxVec == 1) {
+      maxVec = 2;
+    }
+
     if (mask) {
       maxVec = std::min(maxVec, getMaskAlignment(mask));
     }
