--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp
@@ -272,6 +272,12 @@ struct LoadOpConversion : public Convert
         ld(dstsOpr, addrOpr, evictOpr).predicate(pred, "b");
 
       if (other) {
+        if (otherIsSplatConstInt) {
+          for (size_t s = valueElemNBits; s < movWidth; s += valueElemNBits) {
+            splatVal |= splatVal << valueElemNBits;
+          }
+        }
+
         for (size_t ii = 0; ii < nWords; ++ii) {
           // PTX doesn't support mov.u8, so we need to use mov.u16
           PTXInstr &mov =
@@ -292,8 +298,6 @@ struct LoadOpConversion : public Convert
           PTXInstr::Operand *opr{};
 
           if (otherIsSplatConstInt) {
-            for (size_t s = 0; s < 32; s += valueElemNBits)
-              splatVal |= splatVal << valueElemNBits;
             opr = ptxBuilder.newConstantOperand(splatVal);
           } else
             opr = ptxBuilder.newOperand(v, readConstraint);
