diff --git a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
index eec592dee5e5..7939ba627b1d 100644
--- a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
@@ -186,13 +186,15 @@ cc_library(
         "lib/Support/BLAKE3/blake3_portable.c",
     ] + select({
         "@platforms//cpu:aarch64": [
-            "lib/Support/BLAKE3/blake3_neon.c",
+            # TODO(b/234415414): temporary disabled
+            #            "lib/Support/BLAKE3/blake3_neon.c",
         ],
         "@platforms//cpu:x86_64": [
-            "lib/Support/BLAKE3/blake3_sse2_x86-64_unix.S",
-            "lib/Support/BLAKE3/blake3_sse41_x86-64_unix.S",
-            "lib/Support/BLAKE3/blake3_avx2_x86-64_unix.S",
-            "lib/Support/BLAKE3/blake3_avx512_x86-64_unix.S",
+            # TODO(b/234415414): temporary disabled
+            # "lib/Support/BLAKE3/blake3_sse2_x86-64_unix.S",
+            # "lib/Support/BLAKE3/blake3_sse41_x86-64_unix.S",
+            # "lib/Support/BLAKE3/blake3_avx2_x86-64_unix.S",
+            # "lib/Support/BLAKE3/blake3_avx512_x86-64_unix.S",
         ],
         "//conditions:default": [
         ],
@@ -217,14 +219,16 @@ cc_library(
     ],
     copts = llvm_copts,
     defines = select({
-        "@platforms//cpu:aarch64": [
-        ],
+        # TODO(b/234415414): temporary disabled
+        #"@platforms//cpu:aarch64": [
+        #],
         "//conditions:default": [
             "BLAKE3_USE_NEON=0",
         ],
     }) + select({
-        "@platforms//cpu:x86_64": [
-        ],
+        # TODO(b/234415414): temporary disabled
+        # "@platforms//cpu:x86_64": [
+        # ],
         "//conditions:default": [
             "BLAKE3_NO_AVX2",
             "BLAKE3_NO_AVX512",

diff --git a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
index 244e93910880..fd5d24ff7f9a 100644
--- a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
@@ -22,6 +22,7 @@ package(
 licenses(["notice"])
 
 exports_files(["LICENSE.TXT"])
+exports_files(["run_lit.sh"])
 
 filegroup(
     name = "c_headers",
