diff --git a/mlir/lib/Support/TypeID.cpp b/mlir/lib/Support/TypeID.cpp
index 872070609490..ba8ff10a44fc 100644
--- a/mlir/lib/Support/TypeID.cpp
+++ b/mlir/lib/Support/TypeID.cpp
@@ -33,7 +33,7 @@ struct ImplicitTypeIDRegistry {
     // Perform a heuristic check to see if this type is in an anonymous
     // namespace. String equality is not valid for anonymous types, so we try to
     // abort whenever we see them.
-#ifndef NDEBUG
+#if 0 // TODO(b/228157954): Temporarily disabled
 #if defined(_MSC_VER)
     if (typeName.contains("anonymous-namespace")) {
 #else
