diff --git a/utils/bazel/llvm-project-overlay/llvm/config.bzl b/utils/bazel/llvm-project-overlay/llvm/config.bzl
index a8795f1082bb..7aa36b98249a 100644
--- a/utils/bazel/llvm-project-overlay/llvm/config.bzl
+++ b/utils/bazel/llvm-project-overlay/llvm/config.bzl
@@ -50,6 +50,7 @@ linux_defines = posix_defines + [
     "HAVE_MALLINFO=1",
     "HAVE_SBRK=1",
     "HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC=1",
+    "HAVE_GETAUXVAL=1",
 ]

 macos_defines = posix_defines + [
diff --git a/utils/bazel/llvm-project-overlay/llvm/include/llvm/Config/config.h b/utils/bazel/llvm-project-overlay/llvm/include/llvm/Config/config.h
index 93695f8e26d2..31b2a8faded8 100644
--- a/utils/bazel/llvm-project-overlay/llvm/include/llvm/Config/config.h
+++ b/utils/bazel/llvm-project-overlay/llvm/include/llvm/Config/config.h
@@ -296,7 +296,7 @@

 /* HAVE_PROC_PID_RUSAGE defined in Bazel */

-#define HAVE_GETAUXVAL 1
+/* #undef HAVE_GETAUXVAL */

 /* Directly provide definitions here behind platform preprocessor definitions.
  * The preprocessor conditions are sufficient to handle all of the configuration