diff --git a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
index 4e90633b89df..c605f77f7d3b 100644
--- a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
@@ -4101,7 +4101,11 @@ write_file(
 // Provide a weak registration stub in case the real SerializeToCubin is not
 // linked in.
 
+#if defined(_MSC_VER)
+void mlir::registerGpuSerializeToCubinPass() {}
+#else
 __attribute__((weak)) void mlir::registerGpuSerializeToCubinPass() {}
+#endif
 """
     ]
 )
