From 7c0cf32838136734add8ef7a736e7e5fe22988f4 Mon Sep 17 00:00:00 2001
From: Tres Popp <tpopp@google.com>
Date: Thu, 1 Sep 2022 08:19:50 +0200
Subject: [PATCH] bazel fixes for c55b41d5199d2394dd6cdb8f52180d8b81d809d4

---
 .../llvm-project-overlay/mlir/BUILD.bazel     | 23 +++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
index dcd97a9541d3d..c58605965403c 100644
--- a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
@@ -1122,6 +1122,20 @@ gentbl_cc_library(
     name = "AMDGPUIncGen",
     strip_include_prefix = "include",
     tbl_outs = [
+        (
+            [
+                "-gen-attrdef-decls",
+                "-dialect=amdgpu",
+            ],
+            "include/mlir/Dialect/AMDGPU/AMDGPUAttributes.h.inc",
+        ),
+        (
+            [
+                "-gen-attrdef-defs",
+                "-dialect=amdgpu",
+            ],
+            "include/mlir/Dialect/AMDGPU/AMDGPUAttributes.cpp.inc",
+        ),
         (
             [
                 "-gen-dialect-decls",
@@ -1136,6 +1150,14 @@ gentbl_cc_library(
             ],
             "include/mlir/Dialect/AMDGPU/AMDGPUDialect.cpp.inc",
         ),
+        (
+            ["-gen-enum-decls"],
+            "include/mlir/Dialect/AMDGPU/AMDGPUEnums.h.inc",
+        ),
+        (
+            ["-gen-enum-defs"],
+            "include/mlir/Dialect/AMDGPU/AMDGPUEnums.cpp.inc",
+        ),
         (
             ["-gen-op-decls"],
             "include/mlir/Dialect/AMDGPU/AMDGPU.h.inc",
@@ -3781,6 +3803,7 @@ cc_library(
         ":ConversionPassIncGen",
         ":IR",
         ":LLVMCommonConversion",
+        ":LLVMDialect",
         ":Pass",
         ":ROCDLDialect",
         ":Support",
