diff --git a/cc/defs.bzl b/cc/defs.bzl
index 3448e77..9c298ab 100644
--- a/cc/defs.bzl
+++ b/cc/defs.bzl
@@ -30,9 +30,11 @@ load("//cc/toolchains:cc_toolchain_suite.bzl", _cc_toolchain_suite = "cc_toolcha
 load("//cc/toolchains:compiler_flag.bzl", _compiler_flag = "compiler_flag")
 load("//cc/toolchains:fdo_prefetch_hints.bzl", _fdo_prefetch_hints = "fdo_prefetch_hints")
 load("//cc/toolchains:fdo_profile.bzl", _fdo_profile = "fdo_profile")
+load("@com_google_protobuf//bazel:cc_proto_library.bzl", _protobuf_cc_proto_library = "cc_proto_library")
 
 # Rules
 
+cc_proto_library = _protobuf_cc_proto_library
 cc_library = _cc_library
 cc_binary = _cc_binary
 cc_test = _cc_test