diff --git third_party/pybind11_protobuf/BUILD third_party/pybind11_protobuf/BUILD
index b62eb91..b7d1240 100644
--- a/pybind11_protobuf/BUILD
+++ b/pybind11_protobuf/BUILD
@@ -3,8 +3,6 @@
 load("@pybind11_bazel//:build_defs.bzl", "pybind_library")
 load("@bazel_skylib//rules:common_settings.bzl", "bool_flag")

-licenses(["notice"])
-
 pybind_library(
     name = "enum_type_caster",
     hdrs = ["enum_type_caster.h"],