diff --git a/bazel/cython_library.bzl b/bazel/cython_library.bzl
--- a/bazel/cython_library.bzl
+++ b/bazel/cython_library.bzl
@@ -69,5 +69,6 @@ def pyx_library(name, deps = [], py_deps
 
     shared_objects = []
     defines = kwargs.pop("defines", [])
+    copts = kwargs.pop("copts", [])
     for src in pyx_srcs:
         stem = src.split(".")[0]
@@ -80,6 +81,8 @@ def pyx_library(name, deps = [], py_deps
             ],
             defines = defines,
             linkshared = 1,
+            copts = copts,
+            features = kwargs.get("features", []),
             linkopts = select({
                 # The "-undefined dynamic_lookup" flag allows the shared library to use symbols
                 # that are not defined at link time but will be resolved at runtime.
diff --git a/src/python/grpcio/grpc/_cython/BUILD.bazel b/src/python/grpcio/grpc/_cython/BUILD.bazel
--- a/src/python/grpcio/grpc/_cython/BUILD.bazel
+++ b/src/python/grpcio/grpc/_cython/BUILD.bazel
@@ -35,1 +35,4 @@
     defines = ["GRPC_PYTHON_BUILD=1"],
+    # -Wno-unused-result is needed because Cython generated code triggers unused result warnings.
+    copts = ["-Wno-unused-result"],
+    features = ["-layering_check"],
