--- a/bazel/cython_library.bzl
+++ b/bazel/cython_library.bzl
@@ -80,6 +80,7 @@
             ],
             defines = defines,
             linkshared = 1,
+            features = kwargs.get("features", []),
             linkopts = select({
                 # The "-undefined dynamic_lookup" flag allows the shared library to use symbols
                 # that are not defined at link time but will be resolved at runtime.
