diff --git a/python/repositories.bzl b/python/repositories.bzl
index d58feef..4ce550d 100644
--- a/python/repositories.bzl
+++ b/python/repositories.bzl
@@ -64,7 +64,10 @@ def py_repositories():
     )
     http_archive(
         name = "rules_cc",
-        urls = ["https://github.com/bazelbuild/rules_cc/releases/download/0.0.9/rules_cc-0.0.9.tar.gz"],
+        urls = [
+            "https://mirror.bazel.build/github.com/bazelbuild/rules_cc/releases/download/0.0.9/rules_cc-0.0.9.tar.gz",
+            "https://github.com/bazelbuild/rules_cc/releases/download/0.0.9/rules_cc-0.0.9.tar.gz"
+        ],
         sha256 = "2037875b9a4456dce4a79d112a8ae885bbc4aad968e6587dca6e64f3a0900cdf",
         strip_prefix = "rules_cc-0.0.9",
     )
diff --git a/python/versions.bzl b/python/versions.bzl
index fd385cd..9a52e80 100644
--- a/python/versions.bzl
+++ b/python/versions.bzl
@@ -20,7 +20,7 @@ MACOS_NAME = "mac os"
 LINUX_NAME = "linux"
 WINDOWS_NAME = "windows"
 
-DEFAULT_RELEASE_BASE_URL = "https://github.com/indygreg/python-build-standalone/releases/download"
+DEFAULT_RELEASE_BASE_URL = "https://mirror.bazel.build/github.com/indygreg/python-build-standalone/releases/download"
 
 # When updating the versions and releases, run the following command to get
 # the hashes:
@@ -484,6 +484,19 @@ TOOL_VERSIONS = {
         },
         "strip_prefix": "python",
     },
+    "3.13.0": {
+        "url": "20241008/cpython-{python_version}+20241008-{platform}-{build}.tar.gz",
+        "sha256": {
+            "aarch64-apple-darwin": "5d3cb8d7ca4cfbbe7ae1f118f26be112ee417d982fab8c6d85cfd8ccccf70718",
+            "aarch64-unknown-linux-gnu": "c1142af8f2c85923d2ba8201a35b913bb903a5d15f052c38bbecf2f49e2342dc",
+            "ppc64le-unknown-linux-gnu": "1be64a330499fed4e1f864b97eef5445b0e4abc0559ae45df3108981800cf998",
+            "s390x-unknown-linux-gnu": "c0b1cc51426feadaa932fdd9afd9a9af789916e128e48ac8909f9a269bbbd749",
+            "x86_64-apple-darwin": "b58ca12d9ae14bbd79f9e5cf4b748211ff1953e59abeac63b0f4e8e49845669f",
+            "x86_64-pc-windows-msvc": "c7651a7a575104f47c808902b020168057f3ad80f277e54cecfaf79a9ff50e22",
+            "x86_64-unknown-linux-gnu": "455200e1a202e9d9ef4b630c04af701c0a91dcaa6462022efc76893fc762ec95",
+        },
+        "strip_prefix": "python",
+    },
 }
 
 # buildifier: disable=unsorted-dict-items
@@ -493,6 +506,7 @@ MINOR_MAPPING = {
     "3.10": "3.10.14",
     "3.11": "3.11.9",
     "3.12": "3.12.3",
+    "3.13": "3.13.0",
 }
 
 PLATFORMS = {
