diff --git a/BUILD.bazel b/BUILD.bazel
index cc254457..49120384 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -178,6 +178,10 @@ alias(
 cc_library(
     name = "gtest_main",
     srcs = ["googlemock/src/gmock_main.cc"],
+    hdrs = glob([
+        "googletest/include/gtest/*.h",
+        "googlemock/include/gmock/*.h",
+    ]),
     features = select({
         ":windows": ["windows_export_all_symbols"],
         "//conditions:default": [],
