
--- a/third_party/nvidia/lib/Dialect/NVWS/Transforms/InsertTmemAref.cpp
+++ b/third_party/nvidia/lib/Dialect/NVWS/Transforms/InsertTmemAref.cpp
@@ -66,8 +66,8 @@ struct TmemAccessDag {
     SmallVector<std::unique_ptr<Node>> subDags;
     Node(Operation *op, OpOperand *tokOperand,
          std::optional<PartitionId> partitionId, Node *parent)
-        : op(op), tokOperand(tokOperand), partitionId(partitionId),
-          parent(parent), parentDag(nullptr) {}
+        : op(op), tokOperand(tokOperand), parent(parent), parentDag(nullptr),
+          partitionId(partitionId) {}
 
     // ------------------------------------------------------------------------
 
@@ -391,7 +391,7 @@ struct TMEMAref {
   enum Kind { PUT, GET };
 
   TMEMAref(Value aref, Value origBuffer, Value replToken)
-      : aref(aref), origBuffer(origBuffer), replToken(replToken), kind(PUT) {}
+      : origBuffer(origBuffer), aref(aref), replToken(replToken), kind(PUT) {}
 
   void acquire(OpBuilder &b, Location loc,
                std::pair<std::optional<PartitionId>, StageCluster>

