--- a/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertWarpSpecializeToLLVM.cpp
+++ b/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertWarpSpecializeToLLVM.cpp
@@ -77,7 +77,7 @@
       LLVM::NVIDIA::createSyncWarp(b.getLoc(), b);
     } else {
       NVVM::BarrierOp::create(b, b.getLoc(), TypeRange{}, handle,
-                              b.i32_val(numThreads), {}, Value{});
+                              b.i32_val(numThreads));
     }
   }
 

