--- a/test/Conversion/tma_to_llvm.mlir
+++ b/test/Conversion/tma_to_llvm.mlir
@@ -64,7 +64,7 @@
   // CHECK: [[WARP_STRIDE:%.*]] = and i32 [[WARP_STRIDE_TMP]], 768
 
   // CHECK: [[OFFSET0:%.*]] = zext nneg i32 [[WARP_STRIDE]] to i64
-  // CHECK: [[BASEPTR0:%.*]] = getelementptr bfloat, ptr addrspace(3) [[BASE_PTR]], i64 [[OFFSET0]]
+  // CHECK: [[BASEPTR0:%.*]] = getelementptr [2 x i8], ptr addrspace(3) [[BASE_PTR]], i64 [[OFFSET0]]
   // CHECK: "@$0 cp.async.bulk.tensor.2d.tile::gather4.shared::cta.global.mbarrier::complete_tx::bytes [$1], [$2, {$3, $4, $5, $6, $7}], [$8];", "b,r,l,r,r,r,r,r,r"
   // CHECK-SAME: (i1 [[PRED]], ptr addrspace(3) [[BASEPTR0]], ptr nonnull %0, i32 [[Y0]], i32 [[IDX0]], i32 [[IDX1]], i32 [[IDX2]], i32 [[IDX3]], ptr addrspace(3) [[BAR]])
 
@@ -100,7 +100,7 @@
   // CHECK: [[OFFSET0:%.*]] = and i32 [[WARP_STRIDE_TMP]], 1536
 
   // CHECK: zext nneg i32 [[OFFSET0]] to i64
-  // CHECK: [[BASEPTR0:%.*]] = getelementptr bfloat, ptr addrspace(3)
+  // CHECK: [[BASEPTR0:%.*]] = getelementptr [2 x i8], ptr addrspace(3)
   // CHECK: cp.async.bulk.tensor
 
   // CHECK: [[OFFSET1:%.*]] = getelementptr i8, ptr addrspace(3) [[BASEPTR0]], i64 4096

