--- a/include/triton/Dialect/TritonNvidiaGPU/IR/Dialect.h
+++ b/include/triton/Dialect/TritonNvidiaGPU/IR/Dialect.h
@@ -70,6 +70,7 @@ inline bool getModuleTwoCTAs(Operation *op) {
 
 struct TensorMemory : public SideEffects::Resource::Base<TensorMemory> {
   StringRef getName() const final { return "<TensorMemory>"; }
+  Resource* getParent() const override { return nullptr; }
 };
 
 struct TMemAllocation {

