--- a/include/triton/Dialect/Gluon/IR/GluonDialect.td
+++ b/include/triton/Dialect/Gluon/IR/GluonDialect.td
@@ -16,7 +16,6 @@ def Gluon_Dialect : Dialect {
     "mlir::gpu::GPUDialect",
   ];
   let useDefaultAttributePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }
 
 #endif

--- a/include/triton/Dialect/Triton/IR/TritonDialect.td
+++ b/include/triton/Dialect/Triton/IR/TritonDialect.td
@@ -51,7 +51,6 @@ def Triton_Dialect : Dialect {
 
   let hasConstantMaterializer = 1;
   let useDefaultTypePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }
 
 include "triton/Dialect/Triton/IR/TritonTypes.td"

--- a/include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td
+++ b/include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td
@@ -35,7 +35,6 @@ def TritonGPU_Dialect : Dialect {
 
   let useDefaultTypePrinterParser = 1;
   let useDefaultAttributePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }
 
 #endif

--- a/include/triton/Dialect/TritonNvidiaGPU/IR/TritonNvidiaGPUDialect.td
+++ b/include/triton/Dialect/TritonNvidiaGPU/IR/TritonNvidiaGPUDialect.td
@@ -43,7 +43,6 @@ def TritonNvidiaGPU_Dialect : Dialect {
 
   let useDefaultAttributePrinterParser = 1;
   let useDefaultTypePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }
 
 #endif

--- a/third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUDialect.td
+++ b/third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUDialect.td
@@ -37,7 +37,6 @@ def TritonAMDGPU_Dialect : Dialect {
   let dependentDialects = ["triton::TritonDialect"];
 
   let useDefaultAttributePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }
 
 #endif

--- a/third_party/nvidia/include/Dialect/NVWS/IR/NVWSDialect.td
+++ b/third_party/nvidia/include/Dialect/NVWS/IR/NVWSDialect.td
@@ -39,7 +39,6 @@ def NVWS_Dialect : Dialect {
 
   let useDefaultTypePrinterParser = 1;
   let useDefaultAttributePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }
 
 #endif

--- a/third_party/proton/Dialect/include/Dialect/Proton/IR/ProtonDialect.td
+++ b/third_party/proton/Dialect/include/Dialect/Proton/IR/ProtonDialect.td
@@ -14,7 +14,6 @@ def Proton_Dialect : Dialect {
 
   let dependentDialects = [];
 
-  let usePropertiesForAttributes = 1;
 }
 
 #endif

--- a/third_party/proton/Dialect/include/Dialect/ProtonGPU/IR/ProtonGPUDialect.td
+++ b/third_party/proton/Dialect/include/Dialect/ProtonGPU/IR/ProtonGPUDialect.td
@@ -22,7 +22,6 @@ def ProtonGPU_Dialect : Dialect {
 
   let useDefaultTypePrinterParser = 1;
   let useDefaultAttributePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }
 
 #endif  // PROTONGPU_DIALECT

