diff --git a/src/cpu/aarch64/acl_inner_product_utils.cpp b/src/cpu/aarch64/acl_inner_product_utils.cpp
index c8fab86f2..0ebbc1cfa 100644
--- a/src/cpu/aarch64/acl_inner_product_utils.cpp
+++ b/src/cpu/aarch64/acl_inner_product_utils.cpp
@@ -127,7 +127,14 @@ status_t init_conf_ip(acl_ip_conf_t &aip, memory_desc_t &src_md,
     // clang-format on
 
     aip.fc_info.weights_trained_layout = wei_layout;
-    if (is_2d && wei_tag != src_tag) { aip.fc_info.transpose_weights = false; }
+    if (is_2d && wei_tag != src_tag) {
+        aip.fc_info.transpose_weights = false;
+
+        if(prop_kind == dnnl_forward_training) {
+            aip.wei_info.set_are_values_constant(false);
+            aip.fc_info.are_weights_reshaped = true;
+      }
+    }
 
     // Either activation or sum is supported as post-op at the moment
     aip.fc_info.activation_info = acl_common_utils::get_acl_act(attr);
