diff --git a/runtime/src/kmp_settings.cpp b/runtime/src/kmp_settings.cpp
index 692ca26..b9dca42 100644
--- a/runtime/src/kmp_settings.cpp
+++ b/runtime/src/kmp_settings.cpp
@@ -806,6 +806,7 @@ static void __kmp_stg_print_wait_policy(kmp_str_buf_t *buffer, char const *name,
     case library_throughput: {
       value = "PASSIVE";
     } break;
+    default : { } break;
     }
   } else {
     switch (__kmp_library) {
@@ -818,6 +819,7 @@ static void __kmp_stg_print_wait_policy(kmp_str_buf_t *buffer, char const *name,
     case library_throughput: {
       value = "throughput";
     } break;
+    default : { } break;
     }
   }
   if (value != NULL) {
@@ -2428,6 +2430,7 @@ static void __kmp_stg_print_affinity(kmp_str_buf_t *buffer, char const *name,
       __kmp_str_buf_print(buffer, "%s", "granularity=group,");
       break;
 #endif /* KMP_GROUP_AFFINITY */
+    default : break;
     }
   }
   if (!KMP_AFFINITY_CAPABLE()) {
@@ -3819,6 +3822,8 @@ static void __kmp_stg_print_omp_schedule(kmp_str_buf_t *buffer,
     case kmp_sch_auto:
       __kmp_str_buf_print(buffer, "%s,%d'\n", "auto", __kmp_chunk);
       break;
+    default:
+      break;
     }
   } else {
     switch (sched) {
@@ -3844,6 +3849,8 @@ static void __kmp_stg_print_omp_schedule(kmp_str_buf_t *buffer,
     case kmp_sch_auto:
       __kmp_str_buf_print(buffer, "%s'\n", "auto");
       break;
+    default:
+      break;
     }
   }
 } // __kmp_stg_print_omp_schedule
