diff --git a/xprof/convert/BUILD b/xprof/convert/BUILD
--- a/xprof/convert/BUILD
+++ b/xprof/convert/BUILD
@@ -1975,8 +1975,8 @@
     visibility = ["//visibility:private"],
     deps = [
         ":storage_client_interface",
-        "@com_github_googlecloudplatform_google_cloud_cpp//:common",
-        "@com_github_googlecloudplatform_google_cloud_cpp//:storage",
+        "@com_github_googlecloudplatform_google_cloud_cpp//google/cloud:google_cloud_cpp_common",
+        "@com_github_googlecloudplatform_google_cloud_cpp//google/cloud/storage:storage_client",
         "@com_google_absl//absl/base:no_destructor",
         "@com_google_absl//absl/log",
         "@com_google_absl//absl/status",
