diff --git a/shardy/dialect/sdy/transforms/propagation/debugging/source_sharding.cc b/shardy/dialect/sdy/transforms/propagation/debugging/source_sharding.cc
index e1ff855..361b1a6 100644
--- a/shardy/dialect/sdy/transforms/propagation/debugging/source_sharding.cc
+++ b/shardy/dialect/sdy/transforms/propagation/debugging/source_sharding.cc
@@ -144,7 +144,7 @@ void saveShardingOriginsOnModule(ModuleOp moduleOp,
                            DictionaryAttr::get(context, entries));
     }
     TypeSwitch<Operation*, void>(owningOp)
-        .Case<func::FuncOp>([&](func::FuncOp funcOp) {
+        .Case<func::FuncOp>([&, value = value](func::FuncOp funcOp) {
           funcOp.setArgAttr(cast<BlockArgument>(value).getArgNumber(),
                             kOriginShardingAttr,
                             DictionaryAttr::get(context, entries));
