diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 0ca7adb..fae3723 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "cea799afc632879f7d08157108a60c67317829a2"
-    LLVM_SHA256 = "18e5c571e267e0d36087450d4db38015b10e634e51559bb9e7256f57088c43df"
+    LLVM_COMMIT = "52a02b6d1e0c6b492495ff79a3a06ce93e6180b8"
+    LLVM_SHA256 = "d3f2e83de83b205276ba54a3039d2e02040eb8f4af3c74c4c4debda1c8bf3797"
 
     tf_http_archive(
         name = name,
