diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 3686a35..730590f 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "c23d6df60d62f971d957e730f6fe55ea89541f6b"
-    LLVM_SHA256 = "0802cc4a9f52a0c3694d508a903da619dbe205005f59635d8e167fbe4008670c"
+    LLVM_COMMIT = "815b0046b8998e3b9e7a964608c725eec8e85d8a"
+    LLVM_SHA256 = "a192d9c643e502713f4a74c9492f880d073dda753a6447d84d9d917b2676cf59"
 
     tf_http_archive(
         name = name,
