diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 2f731c7..40853aa 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "36adf8ecedb64047021265a1e1730773d3b3a9e8"
-    LLVM_SHA256 = "7baedfc21f67f64f054482cbe77cb3049cd4428187cd45799e10ff8eb03dc9f6"
+    LLVM_COMMIT = "a16164d0c258b284117f8c150efaef891e9aa057"
+    LLVM_SHA256 = "9bf678bbf3e352fca29b3312c00cfb6fb255287c6481f21cb86fe917c512b0bc"
 
     tf_http_archive(
         name = name,
