diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 6429d9b..7b1f825 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "42d641ef5cc4bd82f98ef9959a593ca6db66d75d"
-    LLVM_SHA256 = "ec368e9c3b1e1c5eb646c21da65bb54a53060b417e61f2451f3917b35d743abd"
+    LLVM_COMMIT = "fdce0bfb7f84dc3a29acaefe04ee2f3d75d52c46"
+    LLVM_SHA256 = "8bd199d2946e86883c31b322c88b33fb6a871b03f2517f865e6e64acdec5b3a5"
 
     tf_http_archive(
         name = name,
