diff --git i/third_party/llvm/workspace.bzl w/third_party/llvm/workspace.bzl
index 6429d9b..33ba354 100644
--- i/third_party/llvm/workspace.bzl
+++ w/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "42d641ef5cc4bd82f98ef9959a593ca6db66d75d"
-    LLVM_SHA256 = "ec368e9c3b1e1c5eb646c21da65bb54a53060b417e61f2451f3917b35d743abd"
+    LLVM_COMMIT = "e2c74aa535752cd6cf098731608d26275d1e40ac"
+    LLVM_SHA256 = "e8dc2617bc696b9d87542606b889346a33d6f5cb39768f1b1bd2cbdf7fc35936"
 
     tf_http_archive(
         name = name,
