diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index a41bd24..c21b314 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "23309d7d9553af69b2912a159bc2e488acf69255"
-    LLVM_SHA256 = "d2f80fd052b6326eb6143e0c655335fa50617d66f0c8293dc68a9ac9cb42308a"
+    LLVM_COMMIT = "efcfa6e711689ada546c323316145ecd749d380a"
+    LLVM_SHA256 = "81c9afd66109be46e60ff335c51f2ce9dfbf3998dc87497209bf7cfa6d4f2b22"
 
     tf_http_archive(
         name = name,
