diff --git a/src/BUILD.bazel b/src/BUILD.bazel
index bf71e534e2..22377f1a32 100644
--- a/src/BUILD.bazel
+++ b/src/BUILD.bazel
@@ -971,7 +971,6 @@ filegroup(
 	"runtime/NEON/functions/NETranspose.cpp",
 	"runtime/NEON/functions/NEUnstack.cpp",
 	"runtime/NEON/functions/NEWinogradConvolutionLayer.cpp",
-	"runtime/OMP/OMPScheduler.cpp",
 	"runtime/OffsetLifetimeManager.cpp",
 	"runtime/OffsetMemoryPool.cpp",
 	"runtime/OperatorTensor.cpp",
@@ -984,6 +983,10 @@ filegroup(
 	"runtime/Tensor.cpp",
 	"runtime/TensorAllocator.cpp",
 	"runtime/Utils.cpp"]  +
+    select({
+        "//:openmp_flag": ["runtime/OMP/OMPScheduler.cpp"],
+        "//conditions:default": [],
+    }) +
     glob(["**/*.h",
     "**/*.hpp",
     "**/*.inl"]),
