diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index ad5fa7f9..facd52f4 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -19,8 +19,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "fdb39b1112a6db0343de169b79820768bd16c7d9"
-    LLVM_SHA256 = "15b2d88aa699b4c1d47b489fc0d8a4c389deee53d93c4a985e4c8dc251397da3"
+    LLVM_COMMIT = "6df891b916749d7dc92924659db9ffba8ac64c66"
+    LLVM_SHA256 = "948d2e0c50993af5cc2c6122666a58a2033222743c77878f37551e8e4756cca9"
 
     tf_http_archive(
         name = name,
