diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index dadb98a..d345d6f 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "b35f6e26a5015db32a70048159425aee9e850519"
-    LLVM_SHA256 = "d3673b71d436df48f969835515950c7aecf558c9411bb5055ddb481a834b4795"
+    LLVM_COMMIT = "52ed6791f87a3ef862f555f84ba88a7cdf8fe461"
+    LLVM_SHA256 = "5f4230b06dd2ff977919f26e2deb0b82da00f0a3265f60ac206743169693e933"
 
     tf_http_archive(
         name = name,
