diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 6d5c0d5..c560137 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "1b03747ed85cd4a6573b728674e88f4bd3fa844d"
-    LLVM_SHA256 = "58d57df317c6485b543e2a02ab9d2c1c6148b8f9bc0860741dd558b70de1a787"
+    LLVM_COMMIT = "2af2634c64b1c6f34c4905b24863b930effe3b9c"
+    LLVM_SHA256 = "ec58badb7c10cb9ca8ca064614503adf7f509ebac28a4268f8050ec87f43e90d"
 
     tf_http_archive(
         name = name,
