diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 789c1df..3efaa6c 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "59f57be94f38758616b1339b293b43af845571af"
-    LLVM_SHA256 = "a44b5354b160850cffef1282d912a1bb797864873253680cba82fe210e8ad28e"
+    LLVM_COMMIT = "b22cc5a650deb729f4a1f18206e14971616d07ee"
+    LLVM_SHA256 = "57cca992593ba899d9fb93d6e8c49261043c0ad7b677fad5ed22784ca4ae75a0"
 
     tf_http_archive(
         name = name,
