diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 33b3b01..fc9e8cf 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "f739aa4004165dc64d3a1f418d5ad3c84886f01a"
-    LLVM_SHA256 = "85da134e7ba044ef50ebc009d1a57a87fb0e2db208a04650ef2fe564e9564aa7"
+    LLVM_COMMIT = "3ef78188d0d39cd00429f77f1b300be9bdf85770"
+    LLVM_SHA256 = "3c8263272c72c53bdc8db53b8b2c4d3731cc1661964b9923a7c86dca0f8bd6f2"
 
     tf_http_archive(
         name = name,
