diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 50fd796..a60a18e 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "56ebe64ce69adde8b10793de7aa571df00c75e08"
-    LLVM_SHA256 = "2ef20f6c3bfc2a2e9479cda7b9f34e60dabc8012bd5b06dced883a1be2b0bff4"
+    LLVM_COMMIT = "eed98e1493414ae9c30596b1eeb8f4a9b260e42a"
+    LLVM_SHA256 = "006354b825f373aa49f1c42ca120a1edeaf30c411a49f96b6f79f9cf4a4fb81d"
 
     tf_http_archive(
         name = name,
