diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index abe15ef..9010b35 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "df0864e761107b07e38f5503e0cbee0cebb4c5e8"
-    LLVM_SHA256 = "5bfcb7306d9d40f420862ace1f7ad3f01979facfb16ffd1fc80b6d91e92019fa"
+    LLVM_COMMIT = "0de1e3e787c6d78b87ccb865ba2f2daf8d8e4ecc"
+    LLVM_SHA256 = "5b7678e12f4c5fddf2b8c99d660cd600c076da6de76bb41136510ae816fd9d95"
 
     tf_http_archive(
         name = name,
