diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 08bc617..4957ce4 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "f5dab90875c4435890a19410fcb4bcd8a96357b2"
-    LLVM_SHA256 = "1d5ac81b862040b4e9a87fd0d70f4fa182199ca211654c9723300b6003079876"
+    LLVM_COMMIT = "125948a058dcd89b7fe377872a5fc1a7f9d34e70"
+    LLVM_SHA256 = "8f79c221169dec52116dea2202be64aa653584f98b03a715c30d74ea9141328b"
 
     tf_http_archive(
         name = name,
