diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 98e83ed..a12f943 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "704240125ddf17b9b4995871af3759c742a202ba"
-    LLVM_SHA256 = "0354f1e0c3842bb9f36e590e17bbb91018b014bd5c72736dff0b14d1b3f256e8"
+    LLVM_COMMIT = "f58aa0ec8b3523f3bdaa73964b809d6d54c42768"
+    LLVM_SHA256 = "8c80dd84c614b0a2f05bbc669408e1978d7133442ca1dae8af9bda8a81c0a90a"
 
     tf_http_archive(
         name = name,
