diff --git a/shardy/dialect/sdy/transforms/propagation/op_sharding_rule_registry.cc b/shardy/dialect/sdy/transforms/propagation/op_sharding_rule_registry.cc
index 7d5204f8..c0f05d87 100644
--- a/shardy/dialect/sdy/transforms/propagation/op_sharding_rule_registry.cc
+++ b/shardy/dialect/sdy/transforms/propagation/op_sharding_rule_registry.cc
@@ -666,7 +666,7 @@ OpShardingRuleAttr createOpShardingRule(Operation* op,
                          inShapeRhs[nonBatchDim1], FactorType::kNeedReplication)
               .build();
         }
-        if (callTargetName == "mhlo.topk" || callTargetName == "TopK") {
+        if (callTargetName == "mhlo.topk") {
           assert(customCall.getNumOperands() == 1 &&
                  customCall.getNumResults() == 2);
           // See `jax.lax.top_k` for more information.
diff --git a/shardy/dialect/sdy/transforms/propagation/test/op_sharding_rule_registry.mlir b/shardy/dialect/sdy/transforms/propagation/test/op_sharding_rule_registry.mlir
index 4679d1a6..7e598183 100644
--- a/shardy/dialect/sdy/transforms/propagation/test/op_sharding_rule_registry.mlir
+++ b/shardy/dialect/sdy/transforms/propagation/test/op_sharding_rule_registry.mlir
@@ -475,18 +475,6 @@ func.func @custom_call_top2_of_2d(%arg0: tensor<16x8xf32>) -> (tensor<16x2xf32>,
   return %0#0, %0#1 : tensor<16x2xf32>, tensor<16x2xi32>
 }
 
-// CHECK-LABEL: func @custom_call_topk_target_name
-func.func @custom_call_topk_target_name(%arg0: tensor<16x8xf32>) -> (tensor<16x2xf32>, tensor<16x2xi32>) {
-  // CHECK: sdy.sharding_rule = #sdy.op_sharding_rule<([i, j])->([i, j], [i, j]) {i=16, j=8} need_replication={j} blocked_propagation={j}>
-  %0:2 = stablehlo.custom_call @TopK(%arg0) {
-    mhlo.attributes = {
-        k = 2 : i64,
-        largest = true},
-    mhlo.version = 1 : i64}
-    : (tensor<16x8xf32>) -> (tensor<16x2xf32>, tensor<16x2xi32>)
-  return %0#0, %0#1 : tensor<16x2xf32>, tensor<16x2xi32>
-}
-
 // CHECK-LABEL: func @custom_call_approx_topk
 func.func @custom_call_approx_topk(%arg0: tensor<16x4xf32>, %arg1: tensor<16x4xf32>, %arg2: tensor<f32>, %arg3: tensor<i32>) -> (tensor<16x2xf32>, tensor<16x2xf32>) {
   // CHECK: sdy.sharding_rule = #sdy.op_sharding_rule<([i, j], [i, j], [], [])->([i, k], [i, k]) {i=16, j=4, k=2} need_replication={k} blocked_propagation={k}>}
diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 62fd3f74..c9d59cc7 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -19,8 +19,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "0038be81b8b5361112faf82e69d4590364ee81b9"
-    LLVM_SHA256 = "6675de5b328d62a6bbbef606720ab110d32c55e4645d4fa503e1cfab15bfe878"
+    LLVM_COMMIT = "a22291abb68744cb0b70978eb506a7c5592a3309"
+    LLVM_SHA256 = "cf19b82f0dfbef1e62afe520ba4ee6de4f4823524d7804490e88d815f822c8e9"
 
     tf_http_archive(
         name = name,
