diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 3efaa6c..ce1b210 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "b22cc5a650deb729f4a1f18206e14971616d07ee"
-    LLVM_SHA256 = "57cca992593ba899d9fb93d6e8c49261043c0ad7b677fad5ed22784ca4ae75a0"
+    LLVM_COMMIT = "fe1c4f0106fe4fd6d61c38ba46e71fda8f4d1573"
+    LLVM_SHA256 = "09d358fee9cc681c24e130fa022aaee43e12aafba9124bd5ceab0bec13b37d25"
 
     tf_http_archive(
         name = name,
