diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index ad42b42..63355e1 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "03730cdd3d10c5270fe436777a37d50b0838a3bf"
-    LLVM_SHA256 = "54d843249c75b200f7bf9b7947079fe16fa0b657c4aee4abdde4ac05a9cd5f84"
+    LLVM_COMMIT = "b3134fa2338388adf8cfb2d77339d0b042eab9f6"
+    LLVM_SHA256 = "b6024606092290b0838735c26ad1c5c239b3e931136b420af8680e3a1156e759"
 
     tf_http_archive(
         name = name,
