diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 8b8c2fe..31578bf 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "9c80eb7c83c6471d4126ef46f85bf673787de521"
-    LLVM_SHA256 = "0a00f1377852ce42e2c7ee8cd376d037b8932ed362cd2ae423a52d291314b213"
+    LLVM_COMMIT = "12bcea3292a1559ecad549b5d34c8abcf19f2626"
+    LLVM_SHA256 = "1ce4ec480f47e8b46cea858a652d18dff18844c821d18836c41edbf36151c209"
 
     tf_http_archive(
         name = name,
