diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index ecc2de4..712becc 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "0e3ba99ad65f7025d37c857f9b587b767f7709e7"
-    LLVM_SHA256 = "0c39257da8ec749025e33a1ec2439b6d080eb701947e31b1583c83c7bdee0661"
+    LLVM_COMMIT = "f8cc509b69cc64a6973990f9f48074d211534509"
+    LLVM_SHA256 = "fd64277435142d8991e9cdd84f1f05b8ae4abc1215a2f47e95ba19eccef07de7"
 
     tf_http_archive(
         name = name,
