diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index fae3723..c34853a 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "52a02b6d1e0c6b492495ff79a3a06ce93e6180b8"
-    LLVM_SHA256 = "d3f2e83de83b205276ba54a3039d2e02040eb8f4af3c74c4c4debda1c8bf3797"
+    LLVM_COMMIT = "1c583c19bb7914a2686e245b7e1d14f82fe454eb"
+    LLVM_SHA256 = "28e8b1d1f4af27bf192f63f4294d024e7c30e3e6254d57c0cac44108be347448"
 
     tf_http_archive(
         name = name,
