diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 7c3fc0d..0b17c72 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "e1286d963eeba241e2b1b67f3b1e716dfab83d95"
-    LLVM_SHA256 = "b3275fae17852c429b34b8295eb5f387b12d002de98bc1cc0e68586edcf0f15c"
+    LLVM_COMMIT = "34875331d956862d6f5b750c0bcf026d40564671"
+    LLVM_SHA256 = "d5d86e6d42c4f3a36d558c1b171da180312b1f30ce86a7332cd595cbc16c7a76"
 
     tf_http_archive(
         name = name,
