diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index f3ef519..9c7abe8 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "1061c6da53a7a2c36b6ce8f7499b850212483562"
-    LLVM_SHA256 = "295b804a106f75f1a3c026c3157a64d09e9adea5fed2c912d73e5842e436dd3c"
+    LLVM_COMMIT = "eaa95a1c2bd38332c1a4e634595f29d22b28ffea"
+    LLVM_SHA256 = "c9a50bb272d70f0ee360642d035761b27298afe78adf5c8a30dca50529ac21d6"
 
     tf_http_archive(
         name = name,
