diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index daf3e74..807db4f 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "0c2701fe7fa002e1befc5f86c268a7964f96d286"
-    LLVM_SHA256 = "50b0cb64adf161b4441ee32453f2eb9a015263e4eb3c036548f0207cbe3d0707"
+    LLVM_COMMIT = "0bd2f12753604cd072ae0935820ba9a23bb17ccc"
+    LLVM_SHA256 = "e60e866c649331e41907124f8b3537ecea10acf49da7c5a344d78089df492f0e"
 
     tf_http_archive(
         name = name,
