diff -ur a/BUILD b/BUILD
--- a/BUILD	2017-10-10 15:50:34.000000000 +0000
+++ b/BUILD	2017-10-15 21:19:02.057606476 +0000
@@ -63,6 +63,7 @@
     "-Wwrite-strings",
     "-Wshadow",
     "-fno-common",
+    "-Wno-uninitialized",
 
     # Modern build environments should be able to set this to use atomic
     # operations for reference counting rather than locks. However, it's
diff -ur a/src/include/openssl/base.h b/src/include/openssl/base.h
--- a/src/include/openssl/base.h	2017-10-10 15:50:34.000000000 +0000
+++ b/src/include/openssl/base.h	2017-10-15 19:49:38.182154627 +0000
@@ -106,6 +106,8 @@
 #define OPENSSL_PNACL
 #elif defined(__myriad2__)
 #define OPENSSL_32_BIT
+#elif defined(__s390x__)
+#define OPENSSL_64_BIT
 #else
 // Note BoringSSL only supports standard 32-bit and 64-bit two's-complement,
 // little-endian architectures. Functions will not produce the correct answer
