diff -ur a/src/include/openssl/base.h b/src/include/openssl/base.h
--- a/src/include/openssl/base.h	2017-10-17 21:30:48.000000000 +0000
+++ b/src/include/openssl/base.h	2017-10-18 04:07:23.590036926 +0000
@@ -106,6 +106,8 @@
 #define OPENSSL_PNACL
 #elif defined(__myriad2__)
 #define OPENSSL_32_BIT
+#elif defined(__s390x__)
+#define OPENSSL_64_BIT
 #else
 // Note BoringSSL only supports standard 32-bit and 64-bit two's-complement,
 // little-endian architectures. Functions will not produce the correct answer
