
--- a/lib/Analysis/AxisInfo.cpp	2024-10-01 12:24:54.000000000 -0700
+++ b/lib/Analysis/AxisInfo.cpp	2024-10-15 05:20:45.000000000 -0700
@@ -1079,8 +1079,8 @@
 
 void AxisInfoAnalysis::visitForOpInductionVar(
     scf::ForOp op, ArrayRef<dataflow::Lattice<AxisInfo> *> argLattices) {
-  auto lb = getLatticeElementFor(op, op.getLowerBound())->getValue();
-  auto step = getLatticeElementFor(op, op.getStep())->getValue();
+  auto lb = getLatticeElementFor(getProgramPointAfter(op), op.getLowerBound())->getValue();
+  auto step = getLatticeElementFor(getProgramPointAfter(op), op.getStep())->getValue();
 
   AxisInfo::DimVectorT knownContiguity(1, 1);
   AxisInfo::DimVectorT knownDivisibility(1, 1);

--- a/lib/Analysis/Utility.cpp	2024-10-02 02:26:53.000000000 -0700
+++ b/lib/Analysis/Utility.cpp	2024-10-15 05:20:45.000000000 -0700
@@ -826,15 +826,15 @@
 
   LogicalResult initialize(Operation *top) override {
     WalkResult result = top->walk([&](Operation *op) {
-      if (failed(visit(op)))
+      if (failed(visit(getProgramPointAfter(op))))
         return WalkResult::interrupt();
       return WalkResult::advance();
     });
     return success(!result.wasInterrupted());
   }
 
-  LogicalResult visit(ProgramPoint point) override {
-    Operation *op = point.get<Operation *>();
+  LogicalResult visit(ProgramPoint* point) override {
+    Operation *op = point->getPrevOp();
     Attribute value;
     if (matchPattern(op, m_Constant(&value))) {
       auto *constant = getOrCreate<dataflow::Lattice<dataflow::ConstantValue>>(
