
--- a/BUILD	2025-02-07 01:23:11.000000000 -0800
+++ b/BUILD	2025-02-17 01:49:18.000000000 -0800
@@ -639,6 +639,7 @@
         "@llvm-project//mlir:Analysis",
         "@llvm-project//mlir:ControlFlowDialect",
         "@llvm-project//mlir:DataLayoutInterfaces",
+        "@llvm-project//mlir:FuncToLLVM",
         "@llvm-project//mlir:FunctionInterfaces",
         "@llvm-project//mlir:GPUDialect",
         "@llvm-project//mlir:IR",
