
--- a/include/triton/Dialect/Triton/IR/TritonOps.td	2025-02-03 07:46:30.000000000 -0800
+++ b/include/triton/Dialect/Triton/IR/TritonOps.td	2025-02-03 17:40:44.000000000 -0800
@@ -1106,7 +1106,12 @@
     MutableOperandRange getArgOperandsMutable() {
       return getOperandsMutable();
     }
-
+    Attribute removeArgAttrsAttr() { return nullptr; }
+    Attribute removeResAttrsAttr() { return nullptr; }
+    ArrayAttr getArgAttrsAttr() { return nullptr; }
+    ArrayAttr getResAttrsAttr() { return nullptr; }
+    void setArgAttrsAttr(ArrayAttr) { return; }
+    void setResAttrsAttr(ArrayAttr) { return; }
   }];
 
   let assemblyFormat = [{

--- a/lib/Dialect/Triton/IR/Ops.cpp	2025-01-31 01:23:09.000000000 -0800
+++ b/lib/Dialect/Triton/IR/Ops.cpp	2025-02-03 17:40:45.000000000 -0800
@@ -899,7 +899,7 @@
   if (argAttrs.empty())
     return;
   assert(type.getNumInputs() == argAttrs.size());
-  function_interface_impl::addArgAndResultAttrs(
+  call_interface_impl::addArgAndResultAttrs(
       builder, state, argAttrs, /*resultAttrs=*/std::nullopt,
       getArgAttrsAttrName(state.name), getResAttrsAttrName(state.name));
 }
