
--- a/test/TritonGPU/amd/accelerate-amd-matmul-mfma.mlir	2025-01-21 05:40:49.000000000 -0800
+++ b/test/TritonGPU/amd/accelerate-amd-matmul-mfma.mlir	2025-02-19 08:53:36.000000000 -0800
@@ -1,5 +1,5 @@
-// RUN: triton-opt %s -split-input-file --tritonamdgpu-accelerate-matmul='arch-generation-name=gfx940 matrix-instruction-size=0' | FileCheck %s --check-prefixes MFMA0,CHECK
-// RUN: triton-opt %s -split-input-file --tritonamdgpu-accelerate-matmul='arch-generation-name=gfx940 matrix-instruction-size=16' | FileCheck %s --check-prefixes MFMA16,CHECK
+// RUN: triton-opt %s -split-input-file --tritonamdgpu-accelerate-matmul='arch-generation-name=gfx942 matrix-instruction-size=0' | FileCheck %s --check-prefixes MFMA0,CHECK
+// RUN: triton-opt %s -split-input-file --tritonamdgpu-accelerate-matmul='arch-generation-name=gfx942 matrix-instruction-size=16' | FileCheck %s --check-prefixes MFMA16,CHECK
 
 #blocked = #ttg.blocked<{sizePerThread = [4, 4], threadsPerWarp = [8, 8], warpsPerCTA = [2, 4], order = [1, 0]}>
 // CHECK-LABEL: mfma_dot_fp8e5m2

--- a/test/TritonGPU/amd/amd-convert-buffer-ops.mlir	2025-01-31 01:23:09.000000000 -0800
+++ b/test/TritonGPU/amd/amd-convert-buffer-ops.mlir	2025-02-19 08:53:36.000000000 -0800
@@ -1,4 +1,4 @@
-// RUN: triton-opt %s -split-input-file --tritonamdgpu-convert-buffer-ops='arch-generation-name=gfx940'| FileCheck %s
+// RUN: triton-opt %s -split-input-file --tritonamdgpu-convert-buffer-ops='arch-generation-name=gfx942'| FileCheck %s
 
 #blocked0 = #ttg.blocked<{sizePerThread = [8], threadsPerWarp = [32], warpsPerCTA = [1], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>
 module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 1 : i32} {

--- a/third_party/amd/lib/TritonAMDGPUToLLVM/TargetUtils.cpp	2025-01-21 05:40:49.000000000 -0800
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/TargetUtils.cpp	2025-02-19 08:05:12.000000000 -0800
@@ -13,8 +13,6 @@
   switch (kind) {
   case llvm::AMDGPU::GK_GFX950:
   case llvm::AMDGPU::GK_GFX942:
-  case llvm::AMDGPU::GK_GFX941:
-  case llvm::AMDGPU::GK_GFX940:
     return ISAFamily::CDNA3;
   case llvm::AMDGPU::GK_GFX90A:
     return ISAFamily::CDNA2;
