
--- a/third_party/amd/lib/TritonAMDGPUToLLVM/ScalarizePackedFOps.cpp	2025-06-02 05:51:09.000000000 -0700
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/ScalarizePackedFOps.cpp	2025-06-16 12:28:14.000000000 -0700
@@ -1,4 +1,5 @@
 #include "TritonAMDGPUToLLVM/Passes.h"
+#include "llvm/IR/IRBuilder.h"
 #include "llvm/IR/Instructions.h"
 #include "llvm/IR/PatternMatch.h"
 #include "llvm/IR/Verifier.h"
