
--- a/third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp	2025-04-11 01:29:32.000000000 -0700
+++ b/third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp	2025-07-17 16:41:52.000000000 -0700
@@ -471,7 +471,7 @@
 
   explicit TritonAMDGPULowerInstructionSchedHints(StringRef arch,
                                                   int32_t numStages) {
-    this->arch = std::move(arch.str());
+    this->arch = arch.str();
     this->numStages = numStages;
   }
 
@@ -504,7 +504,7 @@
           TritonAMDGPUInsertInstructionSchedHints> {
 
   explicit TritonAMDGPUInsertInstructionSchedHints(StringRef variant) {
-    this->variant = std::move(variant.str());
+    this->variant = variant.str();
   }
 
   void runOnOperation() override {
