diff --git a/BUILD b/BUILD
index 246fe7e5f..6dba44973 100644
--- a/BUILD
+++ b/BUILD
@@ -951,6 +951,7 @@ cc_library(
         ":triton_conversion_triton_to_triton_gpu_passes_inc_gen",
         ":triton_nvidia_gpu_transforms_inc_gen",
         "@llvm-project//mlir:AllPassesAndDialects",
+        "@llvm-project//mlir:RegisterAllPasses",
         "@triton//test:TritonTestAnalysis",
         "@triton//test:TritonTestDialect",
         "@triton//third_party/amd:TritonAMDGPU",
