
--- a/python/src/llvm.cc	2025-08-22 04:02:56.000000000 -0700
+++ b/python/src/llvm.cc	2025-10-23 11:14:07.000000000 -0700
@@ -53,7 +53,6 @@
   bool disableLLVMOpt = mlir::triton::tools::getBoolEnv("DISABLE_LLVM_OPT");
   if (enable_fp_fusion)
     opt.AllowFPOpFusion = llvm::FPOpFusion::Fast;
-  opt.UnsafeFPMath = false;
   opt.NoInfsFPMath = false;
   opt.NoNaNsFPMath = true;
   opt.TrapUnreachable = true;
