
--- a/lib/Conversion/TritonToTritonGPU/RelayoutTritonGPU.cpp	2025-11-12 02:33:41.000000000 -0800
+++ b/lib/Conversion/TritonToTritonGPU/RelayoutTritonGPU.cpp	2025-12-01 08:23:05.000000000 -0800
@@ -115,7 +115,12 @@
         // clang-format on
         >(typeConverter, context);
 
-    if (failed(applyPartialConversion(mod, target, std::move(patterns))))
+    // config
+    ConversionConfig config;
+    config.allowPatternRollback = false;  // not needed and not supported
+
+    if (failed(
+            applyPartialConversion(mod, target, std::move(patterns), config)))
       return signalPassFailure();
   }
 };
