
--- a/test/Conversion/warp_specialize_to_llvm.mlir	2025-10-15 10:11:13.000000000 -0700
+++ b/test/Conversion/warp_specialize_to_llvm.mlir	2025-11-23 12:53:49.000000000 -0800
@@ -653,7 +653,7 @@
   %1 = llvm.getelementptr %0[%arg0] : (!llvm.ptr<3>, i32) -> !llvm.ptr<3>, i32
   %2 = llvm.add %arg0, %arg1 : i32
   %3 = llvm.mul %2, %arg1 : i32
-  %4 = llvm.udiv %2, %3 : i32
+  %4 = llvm.sub %2, %3 : i32
   ttg.warp_specialize(%1, %4) attributes {allocation.offset = 0 : i32, warpGroupStartIds = array<i32: 4>}
   default {
     ttg.warp_yield
@@ -663,9 +663,9 @@
     // CHECK-NEXT: "llvm.nvvm.barrier.cta.sync.all"([[C1]])
     // CHECK-NEXT: [[ADD:%.*]] = llvm.add %arg0, %arg1 : i32
     // CHECK-NEXT: [[MUL:%.*]] = llvm.mul [[ADD]], %arg1 : i32
-    // CHECK-NEXT: [[UDIV:%.*]] = llvm.udiv [[ADD]], [[MUL]] : i32
+    // CHECK-NEXT: [[SUB:%.*]] = llvm.sub [[ADD]], [[MUL]] : i32
     // CHECK-NEXT: [[PTR:%.*]] = llvm.getelementptr [[ADDR]][%arg0]
-    // CHECK-NEXT: "use"([[PTR]], [[UDIV]])
+    // CHECK-NEXT: "use"([[PTR]], [[SUB]])
     "use"(%arg2, %arg3) : (!llvm.ptr<3>, i32) -> ()
     // CHECK-NEXT: "llvm.nvvm.barrier.cta.sync.all"([[C1]])
     ttg.warp_return
