
--- a/lib/Dialect/TritonGPU/Transforms/Utility.cpp	2025-12-11 07:03:23.000000000 -0800
+++ b/lib/Dialect/TritonGPU/Transforms/Utility.cpp	2026-01-13 20:43:19.000000000 -0800
@@ -1363,7 +1363,11 @@
     for (auto yieldOperand : llvm::enumerate(yieldOp->getOperands())) {
       if (aliveValues.contains(yieldOperand.value()))
         continue;
-      if (yieldOperand.value() == block.getArgument(yieldOperand.index() + 1))
+
+      // If yield operand is a block argument or an init argument, don't bother
+      // with it as it will be removed anyway by the ForOp canonicalizations.
+      if (yieldOperand.value() == block.getArgument(yieldOperand.index() + 1) ||
+          yieldOperand.value() == forOp.getInitArgs()[yieldOperand.index()])
         continue;
 
       // The yield operand might live outside the loop, e.g.
@@ -1383,8 +1387,7 @@
       // can still mark the operand as dead. This occurs in the above example
       // when %init is the same as %x.
       if (!forOp->isAncestor(
-              yieldOperand.value().getParentRegion()->getParentOp()) &&
-          yieldOperand.value() != forOp.getInitArgs()[yieldOperand.index()])
+              yieldOperand.value().getParentRegion()->getParentOp()))
         continue;
 
       deadArg.push_back(yieldOperand.index());
@@ -1393,10 +1396,10 @@
       return failure();
     rewriter.modifyOpInPlace(forOp, [&]() {
       // For simplicity we just change the dead yield operand to use the
-      // associated argument and leave the operations and argument removal to
+      // associated init value and leave the operations and argument removal to
       // dead code elimination.
       for (unsigned deadArgIdx : deadArg) {
-        BlockArgument arg = block.getArgument(deadArgIdx + 1);
+        Value arg = forOp.getInitArgs()[deadArgIdx];
         yieldOp.setOperand(deadArgIdx, arg);
       }
     });

--- a/test/TritonGPU/amd/amd-range-analysis.mlir	2025-11-12 02:33:41.000000000 -0800
+++ b/test/TritonGPU/amd/amd-range-analysis.mlir	2026-01-14 16:51:58.000000000 -0800
@@ -1438,8 +1438,7 @@
       %40 = tt.dot %36, %39, %arg8 : tensor<128x32xf16, #ttg.dot_op<{opIdx = 0, parent = #mma, kWidth = 2}>> * tensor<32x128xf16, #ttg.dot_op<{opIdx = 1, parent = #mma, kWidth = 2}>> -> tensor<128x128xf32, #mma>
       %41 = arith.addi %arg9, %c1_i32 : i32
       %42 = arith.cmpi slt, %41, %c1_i32 : i32
-      // expected-remark@+2 {{unsigned : [0, 0] signed : [0, 0]}}
-      // expected-remark@+1 {{non-neg}}
+      // expected-remark@+1 {{unsigned : [0, 4294967295] signed : [-2147483648, 2147483647]}}
       %43 = arith.select %42, %41, %c0_i32 : i32
       %44 = ttg.memdesc_index %10[%43] : !ttg.memdesc<1x128x32xf16, #shared, #smem, mutable> -> !ttg.memdesc<128x32xf16, #shared, #smem, mutable>
       ttg.local_store %35, %44 : tensor<128x32xf16, #blocked1> -> !ttg.memdesc<128x32xf16, #shared, #smem, mutable>
