
--- a/third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TaskIdPropagation.h	2025-07-31 00:13:23.000000000 -0700
+++ b/third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TaskIdPropagation.h	2026-01-03 23:14:00.000000000 -0800
@@ -86,6 +86,9 @@
   void visitBranchOperand(OpOperand &operand) override;
 
   void visitCallOperand(OpOperand &operand) override;
+
+  void visitNonControlFlowArguments(
+      RegionSuccessor& successor, ArrayRef<BlockArgument> arguments) override {}
 
   void setToExitState(TaskIdLattice *lattice) override;
 
