
--- a/python/src/llvm.cc	2026-01-21 01:19:21.000000000 -0800
+++ b/python/src/llvm.cc	2026-02-12 09:27:22.000000000 -0800
@@ -57,7 +57,6 @@
   bool disableLLVMOpt = mlir::triton::tools::getBoolEnv("DISABLE_LLVM_OPT");
   if (enable_fp_fusion)
     opt.AllowFPOpFusion = llvm::FPOpFusion::Fast;
-  opt.NoInfsFPMath = false;
   opt.NoNaNsFPMath = true;
   opt.TrapUnreachable = true;
   opt.MCOptions.AsmVerbose = true;
