#  Bug tracking the issue:
#     https://b/296812125

diff --git triton/lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp triton/lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp
index e6a69042f..655272778 100644
--- triton/lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp
+++ triton/lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp
@@ -183,7 +183,9 @@ public:
     auto newAcc = rewriter.create<triton::gpu::ConvertLayoutOp>(
         oldAcc.getLoc(), newRetType, oldAcc);
     // convert operands
-    int minBitwidth = std::min(computeOrigBitWidth(a), computeOrigBitWidth(b));
+    // TODO(b/296812125): Fix minBitwidth issue upstream and uncomment.
+    // int minBitwidth = std::min(computeOrigBitWidth(a), computeOrigBitWidth(b));
+    int minBitwidth = 0;
     Type minType = IntegerType::get(ctx, minBitwidth);
     // convert A operand
     auto newAEncoding = triton::gpu::DotOperandEncodingAttr::get(
