diff --git a/include/cudnn_backend_base.h b/include/cudnn_backend_base.h
index 1240282..cba52ec 100644
--- a/include/cudnn_backend_base.h
+++ b/include/cudnn_backend_base.h
@@ -24,7 +24,7 @@
 
 #include <ostream>
 
-#include <cudnn.h>
+#include "third_party/gpus/cudnn/cudnn.h"
 
 namespace cudnn_frontend {
 
diff --git a/include/cudnn_frontend_ConvDesc.h b/include/cudnn_frontend_ConvDesc.h
index 6e1d7ab..4deec88 100644
--- a/include/cudnn_frontend_ConvDesc.h
+++ b/include/cudnn_frontend_ConvDesc.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_Engine.h b/include/cudnn_frontend_Engine.h
index b95efb8..867541e 100644
--- a/include/cudnn_frontend_Engine.h
+++ b/include/cudnn_frontend_Engine.h
@@ -30,8 +30,8 @@
 #include <utility>
 #include <vector>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_OperationGraph.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_EngineConfig.h b/include/cudnn_frontend_EngineConfig.h
index 973e777..97f0883 100644
--- a/include/cudnn_frontend_EngineConfig.h
+++ b/include/cudnn_frontend_EngineConfig.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_Engine.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_EngineFallbackList.h b/include/cudnn_frontend_EngineFallbackList.h
index 4d4e5be..6390bc5 100644
--- a/include/cudnn_frontend_EngineFallbackList.h
+++ b/include/cudnn_frontend_EngineFallbackList.h
@@ -22,7 +22,7 @@
 
 #pragma once
 
-#include <cudnn.h>
+#include "third_party/gpus/cudnn/cudnn.h"
 #include <numeric>
 #include "cudnn_frontend_Heuristics.h"
 
diff --git a/include/cudnn_frontend_ExecutionPlan.h b/include/cudnn_frontend_ExecutionPlan.h
index afceeb3..3d426e2 100644
--- a/include/cudnn_frontend_ExecutionPlan.h
+++ b/include/cudnn_frontend_ExecutionPlan.h
@@ -30,8 +30,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_EngineConfig.h"
 #include "cudnn_frontend_Engine.h"
diff --git a/include/cudnn_frontend_Filters.h b/include/cudnn_frontend_Filters.h
index 676f0f2..4d1c020 100644
--- a/include/cudnn_frontend_Filters.h
+++ b/include/cudnn_frontend_Filters.h
@@ -22,7 +22,7 @@
 
 #pragma once
 
-#include <cudnn.h>
+#include "third_party/gpus/cudnn/cudnn.h"
 
 namespace cudnn_frontend {
 
diff --git a/include/cudnn_frontend_Heuristics.h b/include/cudnn_frontend_Heuristics.h
index dda3fb3..3e89237 100644
--- a/include/cudnn_frontend_Heuristics.h
+++ b/include/cudnn_frontend_Heuristics.h
@@ -25,8 +25,8 @@
 #include <vector>
 #include <mutex>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_OperationGraph.h"
 #include "cudnn_frontend_EngineConfig.h"
diff --git a/include/cudnn_frontend_MatMulDesc.h b/include/cudnn_frontend_MatMulDesc.h
index c9258ba..141f2f9 100644
--- a/include/cudnn_frontend_MatMulDesc.h
+++ b/include/cudnn_frontend_MatMulDesc.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_Operation.h b/include/cudnn_frontend_Operation.h
index bf16cfa..f3086e1 100644
--- a/include/cudnn_frontend_Operation.h
+++ b/include/cudnn_frontend_Operation.h
@@ -30,8 +30,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_ConvDesc.h"
 #include "cudnn_frontend_PointWiseDesc.h"
diff --git a/include/cudnn_frontend_OperationGraph.h b/include/cudnn_frontend_OperationGraph.h
index c5e2704..71589b2 100644
--- a/include/cudnn_frontend_OperationGraph.h
+++ b/include/cudnn_frontend_OperationGraph.h
@@ -30,8 +30,8 @@
 #include <utility>
 #include <vector>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_Operation.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_PointWiseDesc.h b/include/cudnn_frontend_PointWiseDesc.h
index afa71ce..56b6507 100644
--- a/include/cudnn_frontend_PointWiseDesc.h
+++ b/include/cudnn_frontend_PointWiseDesc.h
@@ -30,8 +30,8 @@
 #include <utility>
 #include <limits>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_ReductionDesc.h b/include/cudnn_frontend_ReductionDesc.h
index 5df2c5e..419fc93 100644
--- a/include/cudnn_frontend_ReductionDesc.h
+++ b/include/cudnn_frontend_ReductionDesc.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_Resample.h b/include/cudnn_frontend_Resample.h
index 351e2da..b1a1904 100644
--- a/include/cudnn_frontend_Resample.h
+++ b/include/cudnn_frontend_Resample.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_Rng.h b/include/cudnn_frontend_Rng.h
index 9d4e6ca..4224b61 100644
--- a/include/cudnn_frontend_Rng.h
+++ b/include/cudnn_frontend_Rng.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_VariantPack.h b/include/cudnn_frontend_VariantPack.h
index 455ab8b..4173860 100644
--- a/include/cudnn_frontend_VariantPack.h
+++ b/include/cudnn_frontend_VariantPack.h
@@ -30,8 +30,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
