diff --git a/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise.c b/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise.c
index 3e29856..320987b 100644
--- a/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise.c
+++ b/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise.c
@@ -249,8 +249,6 @@ WORD32 xa_nn_conv2d_depthwise_getsize
     XA_NNLIB_CHK_COND((kernel_height <= 0), -1);
     XA_NNLIB_CHK_COND((kernel_width <= 0), -1);
     XA_NNLIB_CHK_COND((channels_multiplier <= 0), -1);
-    XA_NNLIB_CHK_COND((x_stride <= 0 || x_stride > kernel_width), -1);
-    XA_NNLIB_CHK_COND((y_stride <= 0 || y_stride > kernel_height), -1);
     XA_NNLIB_CHK_COND((x_padding < 0), -1);
     XA_NNLIB_CHK_COND((y_padding < 0), -1);
     XA_NNLIB_CHK_COND((output_height <= 0), -1);
diff --git a/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise_sym8sxasym8s.c b/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise_sym8sxasym8s.c
index e719da1..5b7390f 100644
--- a/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise_sym8sxasym8s.c
+++ b/algo/kernels/cnn/hifi4/xa_nn_conv2d_depthwise_sym8sxasym8s.c
@@ -659,7 +659,6 @@ WORD32 xa_nn_conv2d_depthwise_per_chan_sym8sxasym8s
     XA_NNLIB_ARG_CHK_COND((input_channels <= 0), -1);
     XA_NNLIB_ARG_CHK_COND((kernel_height <= 0 || kernel_width <= 0), -1);
     XA_NNLIB_ARG_CHK_COND((kernel_height > input_height), -1);
-    XA_NNLIB_ARG_CHK_COND((kernel_width > input_width), -1);
     XA_NNLIB_ARG_CHK_COND((channels_multiplier <= 0), -1);
     XA_NNLIB_ARG_CHK_COND((y_stride <= 0 || x_stride <= 0), -1);
     XA_NNLIB_ARG_CHK_COND((y_padding < 0 || x_padding < 0), -1);
@@ -671,8 +670,6 @@ WORD32 xa_nn_conv2d_depthwise_per_chan_sym8sxasym8s
     XA_NNLIB_ARG_CHK_COND((inp_data_format != 0 && inp_data_format != 1), -1);
     XA_NNLIB_ARG_CHK_COND((out_data_format != 0), -1);
     /* Implementation dependent checks */
-    XA_NNLIB_ARG_CHK_COND((y_stride > kernel_height), -1);
-    XA_NNLIB_ARG_CHK_COND((x_stride > kernel_width), -1);
 
     if(inp_data_format == 0)
     {
