diff --git a/absl/container/internal/btree.h b/absl/container/internal/btree.h
--- a/absl/container/internal/btree.h
+++ b/absl/container/internal/btree.h
@@ -1119,10 +1119,8 @@ class btree_iterator : private btree_iterator_generation_info {
   using slot_type = typename params_type::slot_type;
 
   // In sets, all iterators are const.
-  using iterator = absl::conditional_t<
-      is_map_container::value,
-      btree_iterator<normal_node, normal_reference, normal_pointer>,
-      btree_iterator<normal_node, const_reference, const_pointer>>;
+  using iterator =
+      btree_iterator<normal_node, normal_reference, normal_pointer>;
   using const_iterator =
       btree_iterator<const_node, const_reference, const_pointer>;
 
