diff --git a/absl/BUILD.bazel b/absl/BUILD.bazel
--- a/absl/BUILD.bazel	(revision fb3621f4f897824c0dbe0615fa94543df6192f30)
+++ b/absl/BUILD.bazel	(date 1741994158088)
@@ -15,11 +15,26 @@
 #

 load("@bazel_skylib//lib:selects.bzl", "selects")
+load("@bazel_skylib//rules:common_settings.bzl", "bool_flag")

 package(default_visibility = ["//visibility:public"])

 licenses(["notice"])

+bool_flag(
+    name = "use_dlls",
+    build_setting_default = False,
+    visibility = ["//visibility:public"],
+)
+
+config_setting(
+    name = "enable_use_dlls",
+    flag_values = {
+        ":use_dlls": "True",
+    },
+    visibility = [":__subpackages__"],
+)
+
 config_setting(
     name = "clang_compiler",
     flag_values = {
diff --git a/absl/copts/configure_copts.bzl b/absl/copts/configure_copts.bzl
--- a/absl/copts/configure_copts.bzl	(revision fb3621f4f897824c0dbe0615fa94543df6192f30)
+++ b/absl/copts/configure_copts.bzl	(date 1741994109164)
@@ -27,6 +27,9 @@
     "//absl:clang_compiler": ABSL_LLVM_FLAGS,
     "//absl:gcc_compiler": ABSL_GCC_FLAGS,
     "//conditions:default": ABSL_GCC_FLAGS,
+}) + select({
+    "//absl:enable_use_dlls": ["/DABSL_BUILD_DLL"],
+    "//conditions:default": []
 })

 ABSL_TEST_COPTS = select({
