diff --git a/absl/base/BUILD.bazel b/absl/base/BUILD.bazel
--- a/absl/base/BUILD.bazel
+++ b/absl/base/BUILD.bazel
@@ -578,12 +578,8 @@ cc_library(
     copts = ABSL_DEFAULT_COPTS,
     linkopts = ABSL_DEFAULT_LINKOPTS,
-    visibility = [
-        "//absl:__subpackages__",
-        "//absl:friends",
-    ],
     deps = [
         ":base",
         ":config",
         ":core_headers",
         ":nullability",
     ],


