diff --git a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
index e83a237a5397..462b6b60f5b2 100644
--- a/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel
@@ -232,19 +232,19 @@ write_file(
 config_setting(
     name = "is_windows_clang_mingw",
     constraint_values = ["@platforms//os:windows"],
-    flag_values = {"@rules_cc//cc/compiler:compiler": "clang"},
+    flag_values = {"@rules_cc//cc/private/toolchain:compiler": "clang"},
 )
 
 config_setting(
     name = "is_windows_clang_cl",
     constraint_values = ["@platforms//os:windows"],
-    flag_values = {"@rules_cc//cc/compiler:compiler": "clang-cl"},
+    flag_values = {"@rules_cc//cc/private/toolchain:compiler": "clang-cl"},
 )
 
 config_setting(
     name = "is_windows_msvc",
     constraint_values = ["@platforms//os:windows"],
-    flag_values = {"@rules_cc//cc/compiler:compiler": "msvc-cl"},
+    flag_values = {"@rules_cc//cc/private/toolchain:compiler": "msvc-cl"},
 )
 
 config_setting(
@@ -253,7 +253,7 @@ config_setting(
         "@platforms//cpu:x86_64",
         "@platforms//os:windows",
     ],
-    flag_values = {"@rules_cc//cc/compiler:compiler": "clang"},
+    flag_values = {"@rules_cc//cc/private/toolchain:compiler": "clang"},
 )
 
 # TODO(zbarsky): Is there a better way to express this?
@@ -327,12 +327,14 @@ cc_library(
             "lib/Support/BLAKE3/blake3_neon.c",
         ],
         ":is_x86_64_windows_clang_mingw": [
-            pattern % "windows_gnu"
-            for pattern in BLAKE3_x86_64_ASM_SOURCE_PATTERNS
+            # TODO(b/234415414): temporary disabled
+            # pattern % "windows_gnu"
+            # for pattern in BLAKE3_x86_64_ASM_SOURCE_PATTERNS
         ],
         ":is_x86_64_non_windows": [
-            pattern % "unix"
-            for pattern in BLAKE3_x86_64_ASM_SOURCE_PATTERNS
+            # TODO(b/234415414): temporary disabled
+            # pattern % "unix"
+            # for pattern in BLAKE3_x86_64_ASM_SOURCE_PATTERNS
         ],
         "//conditions:default": [
         ],
@@ -358,14 +360,16 @@ cc_library(
     ],
     copts = llvm_copts,
     defines = select({
-        "@platforms//cpu:aarch64": [
-        ],
+        # TODO(b/234415414): temporary disabled
+        #"@platforms//cpu:aarch64": [
+        #],
         "//conditions:default": [
             "BLAKE3_USE_NEON=0",
         ],
     }) + select({
-        "@platforms//cpu:x86_64": [
-        ],
+        # TODO(b/234415414): temporary disabled
+        # "@platforms//cpu:x86_64": [
+        # ],
         "//conditions:default": [
             "BLAKE3_NO_AVX2",
             "BLAKE3_NO_AVX512",
