diff --git a/utils/bazel/llvm-project-overlay/llvm/lit_test.bzl b/utils/bazel/llvm-project-overlay/llvm/lit_test.bzl
--- a/utils/bazel/llvm-project-overlay/llvm/lit_test.bzl
+++ b/utils/bazel/llvm-project-overlay/llvm/lit_test.bzl
@@ -204,7 +204,9 @@ def lit_test(

     main = Label("//llvm/utils/lit:lit.py")
     py_test(
-        name = name,
+        # Prevents creation of the files with identical names located in different
+        # directories on Windows platform.
+        name = name.replace("/", "_"),
         srcs = [main],
         main = main,
         args = args + ["-v"] + ["$(rootpath %s)" % src for src in srcs],
