diff --git a/src/nb_type.cpp b/src/nb_type.cpp
--- a/src/nb_type.cpp
+++ b/src/nb_type.cpp
@@ -36,6 +36,11 @@ static PyObject **nb_weaklist_ptr(PyObje
     return weaklistoffset ? (PyObject **) ((uint8_t *) self + weaklistoffset) : nullptr;
 }
 
+static PyGetSetDef inst_getset[] = {
+    { "__dict__", PyObject_GenericGetDict, PyObject_GenericSetDict, nullptr, nullptr },
+    { nullptr, nullptr, nullptr, nullptr, nullptr }
+};
+
 static int inst_clear(PyObject *self) {
     PyObject **dict = nb_dict_ptr(self);
     if (dict)
@@ -923,8 +928,11 @@ PyObject *nb_type_new(const type_init_da
     }
 
     bool has_traverse = false;
-    for (PyType_Slot *ts = slots; ts != s; ++ts)
+    bool has_getset = false;
+    for (PyType_Slot *ts = slots; ts != s; ++ts) {
         has_traverse |= ts->slot == Py_tp_traverse;
+        has_getset |= ts->slot == Py_tp_getset;
+    }
 
     Py_ssize_t dictoffset = 0, weaklistoffset = 0;
     int num_members = 0;
@@ -948,6 +956,10 @@ PyObject *nb_type_new(const type_init_da
             has_traverse = true;
         }
         spec.basicsize = (int) basicsize;
+
+        if (!has_getset) {
+            *s++ = { Py_tp_getset, (void *) inst_getset };
+        }
     }
 
     if (is_weak_referenceable) {
