--- ./absl/time/internal/cctz/BUILD.bazel	2019-09-23 13:20:52.000000000 -0700
+++ ./absl/time/internal/cctz/BUILD.bazel.fixed	2019-09-23 13:20:48.000000000 -0700
@@ -76,15 +76,6 @@
         "include/cctz/time_zone.h",
         "include/cctz/zone_info_source.h",
     ],
-    linkopts = select({
-        ":osx": [
-            "-framework Foundation",
-        ],
-        ":ios": [
-            "-framework Foundation",
-        ],
-        "//conditions:default": [],
-    }),
     visibility = ["//visibility:public"],
     deps = [":civil_time"],
 )

--- ./absl/time/clock.h	2019-08-08 14:19:45.000000000 -0400
+++ ./absl/time/clock.h.fixed	2019-10-03 19:28:14.630080167 -0400
@@ -67,7 +67,7 @@
 void AbslInternalSleepFor(absl::Duration duration);
 }  // extern "C"

-inline void absl::SleepFor(absl::Duration duration) {
+inline void absl::lts_2019_08_08::SleepFor(absl::lts_2019_08_08::Duration duration) {
   AbslInternalSleepFor(duration);
 }
