diff --git a/include/cudnn_frontend.h b/include/cudnn_frontend.h
index e3f1ec8..373e52d 100644
--- a/include/cudnn_frontend.h
+++ b/include/cudnn_frontend.h
@@ -97,7 +97,7 @@
  *      - Simpler samples on how to use the new API.
  */
 
-#include <cudnn.h>
+#include "third_party/gpus/cudnn/cudnn.h"
 
 #include "cudnn_frontend_ConvDesc.h"
 #include "cudnn_frontend_Heuristics.h"
diff --git a/include/cudnn_frontend/backend/backend_descriptor.h b/include/cudnn_frontend/backend/backend_descriptor.h
index 47387a1..3864c1d 100644
--- a/include/cudnn_frontend/backend/backend_descriptor.h
+++ b/include/cudnn_frontend/backend/backend_descriptor.h
@@ -3,7 +3,7 @@
 #include <memory>
 
 #include "../graph_helpers.h"
-#include "cudnn.h"
+#include "third_party/gpus/cudnn/cudnn.h"
 
 namespace cudnn_frontend::detail {
 
diff --git a/include/cudnn_frontend/backend/execution_helpers.h b/include/cudnn_frontend/backend/execution_helpers.h
index 334ffde..d2ca694 100644
--- a/include/cudnn_frontend/backend/execution_helpers.h
+++ b/include/cudnn_frontend/backend/execution_helpers.h
@@ -2,7 +2,7 @@
 
 #include <vector>
 
-#include "cudnn.h"
+#include "third_party/gpus/cudnn/cudnn.h"
 
 #include "backend_descriptor.h"
 
diff --git a/include/cudnn_frontend/backend/plan_helpers.h b/include/cudnn_frontend/backend/plan_helpers.h
index 1fa458d..8c37d10 100644
--- a/include/cudnn_frontend/backend/plan_helpers.h
+++ b/include/cudnn_frontend/backend/plan_helpers.h
@@ -2,7 +2,7 @@
 
 #include <vector>
 
-#include "cudnn.h"
+#include "third_party/gpus/cudnn/cudnn.h"
 
 #include "backend_descriptor.h"
 
