diff --git a/include/cudnn_backend_base.h b/include/cudnn_backend_base.h
index b07b336..3fb06a7 100644
--- a/include/cudnn_backend_base.h
+++ b/include/cudnn_backend_base.h
@@ -22,7 +22,7 @@
 
 #pragma once
 
-#include <cudnn.h>
+#include "third_party/gpus/cudnn/cudnn.h"
 
 namespace cudnn_frontend {
 
diff --git a/include/cudnn_frontend_ConvDesc.h b/include/cudnn_frontend_ConvDesc.h
index d8706d4..6cabf6e 100644
--- a/include/cudnn_frontend_ConvDesc.h
+++ b/include/cudnn_frontend_ConvDesc.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_Engine.h b/include/cudnn_frontend_Engine.h
index 7b3c612..05ee6a6 100644
--- a/include/cudnn_frontend_Engine.h
+++ b/include/cudnn_frontend_Engine.h
@@ -30,8 +30,8 @@
 #include <utility>
 #include <vector>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_OperationGraph.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_EngineConfig.h b/include/cudnn_frontend_EngineConfig.h
index 37b4e99..1afd3cc 100644
--- a/include/cudnn_frontend_EngineConfig.h
+++ b/include/cudnn_frontend_EngineConfig.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_Engine.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_EngineFallbackList.h b/include/cudnn_frontend_EngineFallbackList.h
index 4fadb44..cb2be8a 100644
--- a/include/cudnn_frontend_EngineFallbackList.h
+++ b/include/cudnn_frontend_EngineFallbackList.h
@@ -22,7 +22,7 @@
 
 #pragma once
 
-#include <cudnn.h>
+#include "third_party/gpus/cudnn/cudnn.h"
 #include <numeric>
 
 namespace cudnn_frontend {
diff --git a/include/cudnn_frontend_ExecutionPlan.h b/include/cudnn_frontend_ExecutionPlan.h
index 03be7ec..1e8ea83 100644
--- a/include/cudnn_frontend_ExecutionPlan.h
+++ b/include/cudnn_frontend_ExecutionPlan.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_Engine.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_Filters.h b/include/cudnn_frontend_Filters.h
index c8ca775..5969ec1 100644
--- a/include/cudnn_frontend_Filters.h
+++ b/include/cudnn_frontend_Filters.h
@@ -22,7 +22,7 @@
 
 #pragma once
 
-#include <cudnn.h>
+#include "third_party/gpus/cudnn/cudnn.h"
 
 namespace cudnn_frontend {
 
diff --git a/include/cudnn_frontend_Heuristics.h b/include/cudnn_frontend_Heuristics.h
index 13a2e32..6a74c5f 100644
--- a/include/cudnn_frontend_Heuristics.h
+++ b/include/cudnn_frontend_Heuristics.h
@@ -24,8 +24,8 @@
 
 #include <vector>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_OperationGraph.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_Operation.h b/include/cudnn_frontend_Operation.h
index f79c69c..840076b 100644
--- a/include/cudnn_frontend_Operation.h
+++ b/include/cudnn_frontend_Operation.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_ConvDesc.h"
 #include "cudnn_frontend_PointWiseDesc.h"
diff --git a/include/cudnn_frontend_OperationGraph.h b/include/cudnn_frontend_OperationGraph.h
index f1d5dcc..5edd0e7 100644
--- a/include/cudnn_frontend_OperationGraph.h
+++ b/include/cudnn_frontend_OperationGraph.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_Operation.h"
 #include "cudnn_frontend_utils.h"
diff --git a/include/cudnn_frontend_PointWiseDesc.h b/include/cudnn_frontend_PointWiseDesc.h
index 4715971..b086d23 100644
--- a/include/cudnn_frontend_PointWiseDesc.h
+++ b/include/cudnn_frontend_PointWiseDesc.h
@@ -29,8 +29,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
diff --git a/include/cudnn_frontend_VariantPack.h b/include/cudnn_frontend_VariantPack.h
index ab2aab3..94aae89 100644
--- a/include/cudnn_frontend_VariantPack.h
+++ b/include/cudnn_frontend_VariantPack.h
@@ -30,8 +30,8 @@
 #include <sstream>
 #include <utility>
 
-#include <cudnn.h>
-#include <cudnn_backend.h>
+#include "third_party/gpus/cudnn/cudnn.h"
+#include "third_party/gpus/cudnn/cudnn_backend.h"
 
 #include "cudnn_frontend_utils.h"
 
