diff --git a/mlir/lib/Dialect/GPU/Transforms/AsyncRegionRewriter.cpp b/mlir/lib/Dialect/GPU/Transforms/AsyncRegionRewriter.cpp
index 74b695eec18e..917c27714b4a 100644
--- a/mlir/lib/Dialect/GPU/Transforms/AsyncRegionRewriter.cpp
+++ b/mlir/lib/Dialect/GPU/Transforms/AsyncRegionRewriter.cpp
@@ -108,10 +108,8 @@ struct GpuAsyncRegionPass::Callback {
 // inserts the necessary synchronization (as gpu.wait ops). Assumes sequential
 // execution semantics and that no GPU ops are asynchronous yet.
 void GpuAsyncRegionPass::runOnFunction() {
-  if (getFunction()
-          .getRegion()
-          .walk(Callback{OpBuilder(&getContext())})
-          .wasInterrupted())
+  Callback callback{OpBuilder(&getContext())};
+  if (getFunction().getRegion().walk(callback).wasInterrupted())
     return signalPassFailure();
 }
 
