diff -r -u ./scripts/pnglibconf.h.prebuilt ./scripts/pnglibconf.h.prebuilt
--- ./scripts/pnglibconf.h.prebuilt
+++ ./scripts/pnglibconf.h.prebuilt
@@ -19,6 +19,12 @@
 #define PNG_ALIGNED_MEMORY_SUPPORTED
 /*#undef PNG_ARM_NEON_API_SUPPORTED*/
 /*#undef PNG_ARM_NEON_CHECK_SUPPORTED*/
+
+/* Workaround not having a great build file by forcing
+ * png filter optimization to be disabled on arm */
+#define PNG_ARM_NEON_OPT 0
+
+
 #define PNG_BENIGN_ERRORS_SUPPORTED
 #define PNG_BENIGN_READ_ERRORS_SUPPORTED
 /*#undef PNG_BENIGN_WRITE_ERRORS_SUPPORTED*/
