diff --git a/absl/container/internal/btree.h b/absl/container/internal/btree.h
--- a/absl/container/internal/btree.h
+++ b/absl/container/internal/btree.h
@@ -1115,15 +1115,13 @@
-  using normal_node = typename std::remove_const<Node>::type;
+  using normal_node = std::remove_const_t<Node>;
   using const_node = const Node;
   using normal_pointer = typename params_type::pointer;
   using normal_reference = typename params_type::reference;
   using const_pointer = typename params_type::const_pointer;
   using const_reference = typename params_type::const_reference;
   using slot_type = typename params_type::slot_type;
 
   // In sets, all iterators are const.
-  using iterator = std::conditional_t<
-      is_map_container::value,
-      btree_iterator<normal_node, normal_reference, normal_pointer>,
-      btree_iterator<normal_node, const_reference, const_pointer>>;
+  using iterator =
+      btree_iterator<normal_node, normal_reference, normal_pointer>;
   using const_iterator =
       btree_iterator<const_node, const_reference, const_pointer>;
