diff --git a/BUILD.bazel b/BUILD.bazel
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -555,7 +555,8 @@ proto_lang_toolchain(
         "//:cpp_features_proto",
         "//:descriptor_proto",
     ],
-    command_line = "--cpp_out=$(OUT)",
+    command_line = "--cpp_out=dllexport_decl=PROTOBUF_EXPORT:$(OUT)",
+#    command_line = "--cpp_out=$(OUT)",
     runtime = "//src/google/protobuf",
     visibility = ["//visibility:public"],
 )
diff --git a/python/dist/system_python.bzl b/python/dist/system_python.bzl
--- a/python/dist/system_python.bzl
+++ b/python/dist/system_python.bzl
@@ -73,11 +73,11 @@ load("@bazel_skylib//lib:selects.bzl", "selects")
 load("@bazel_skylib//rules:common_settings.bzl", "string_flag")
-load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+load("@rules_python//python:py_runtime_pair.bzl", "py_runtime_pair")

-cc_library(
-   name = "python_headers",
-   hdrs = glob(["python/**/*.h"], allow_empty = True),
-   includes = ["python"],
-   visibility = ["//visibility:public"],
+alias(
+    name = "python_headers",
+    actual = "@rules_python//python/cc:current_py_cc_headers",
+    visibility = ["//visibility:public"],
 )

 string_flag(
@@ -219,7 +218,7 @@ def _system_python_impl(repository_ctx):
     python3 = repository_ctx.which("python3")
     python_version = _get_python_version(repository_ctx)

-    if path and python_version[0] == "3":
+    if False:
         _populate_package(repository_ctx, path, python3, python_version)
     else:
         # buildifier: disable=print
diff --git a/python/google/protobuf/__init__.py b/python/google/protobuf/__init__.py
--- a/python/google/protobuf/__init__.py
+++ b/python/google/protobuf/__init__.py
@@ -8,3 +8,9 @@
 # Copyright 2007 Google Inc. All Rights Reserved.

 __version__ = '6.31.1'
+
+if __name__ != '__main__':
+  try:
+    __import__('pkg_resources').declare_namespace(__name__)
+  except ImportError:
+    __path__ = __import__('pkgutil').extend_path(__path__, __name__)
diff --git a/src/google/protobuf/arena.cc b/src/google/protobuf/arena.cc
--- a/src/google/protobuf/arena.cc
+++ b/src/google/protobuf/arena.cc
@@ -547,7 +547,7 @@ ThreadSafeArena::ThreadCache& ThreadSafeArena::thread_cache() {
       new internal::ThreadLocalStorage<ThreadCache>();
   return *thread_cache_->Get();
 }
-#elif defined(PROTOBUF_USE_DLLS) && defined(_WIN32)
+#elif defined(_WIN32)
 ThreadSafeArena::ThreadCache& ThreadSafeArena::thread_cache() {
   static PROTOBUF_THREAD_LOCAL ThreadCache thread_cache;
   return thread_cache;
diff --git a/src/google/protobuf/message_lite.h b/src/google/protobuf/message_lite.h
--- a/src/google/protobuf/message_lite.h
+++ b/src/google/protobuf/message_lite.h
@@ -282,7 +282,11 @@ template <typename T>
 using MessageTraits = decltype(MessageTraitsImpl::value<T>);

 struct EnumTraitsImpl {
+#ifdef __CUDACC__
+  struct Undefined{};
+#else
   struct Undefined;
+#endif
   template <typename T>
   static Undefined value;
 };
diff --git a/src/google/protobuf/port_def.inc b/src/google/protobuf/port_def.inc
--- a/src/google/protobuf/port_def.inc
+++ b/src/google/protobuf/port_def.inc
@@ -421,7 +421,7 @@ static_assert(PROTOBUF_ABSL_MIN(20230125, 3),
 #endif

 // Lexan sets both MSV_VER and clang, so handle it with the clang path.
-#if defined(_MSC_VER) && !defined(__clang__)
+#if defined(_MSC_VER)
 // MSVC 17 currently seems to raise an error about constant-initialized pointers.
 # if PROTOBUF_MSC_VER_MIN(1930)
 #  define PROTOBUF_CONSTINIT
diff --git a/src/google/protobuf/thread_safe_arena.h b/src/google/protobuf/thread_safe_arena.h
--- a/src/google/protobuf/thread_safe_arena.h
+++ b/src/google/protobuf/thread_safe_arena.h
@@ -249,7 +249,7 @@ class PROTOBUF_EXPORT ThreadSafeArena {
   // iOS does not support __thread keyword so we use a custom thread local
   // storage class we implemented.
   static ThreadCache& thread_cache();
-#elif defined(PROTOBUF_USE_DLLS) && defined(_WIN32)
+#elif defined(_WIN32)
   // Thread local variables cannot be exposed through MSVC DLL interface but we
   // can wrap them in static functions.
   static ThreadCache& thread_cache();
